<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Part 2 – BIOL90042 R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./ch8_section_3.html" rel="next">
<link href="./ch8_section_1.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-063905fb60d21bb612e30333561bc9be.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-ee22da2db2cf0c267ae220c203eece4d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./chapter_8.html">8 Machine learning</a></li><li class="breadcrumb-item"><a href="./ch8_section_2.html">Part 2</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">BIOL90042 R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Working with data - Part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Working with data - Part 2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Plotting with ggplot2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Putting it all together</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./stats_primer.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Statistics in R Primer</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">RNA-seq - Part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">RNA-seq - Part 2</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./chapter_8.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">8 Machine learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Part 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 3</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./further_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Further reading</span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this section:</h2>
   
  <ul>
  <li><a href="#background" id="toc-background" class="nav-link active" data-scroll-target="#background">Background</a></li>
  <li><a href="#aim" id="toc-aim" class="nav-link" data-scroll-target="#aim">Aim</a></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries">Load libraries</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#set-seed" id="toc-set-seed" class="nav-link" data-scroll-target="#set-seed">Set seed</a></li>
  <li><a href="#format-data" id="toc-format-data" class="nav-link" data-scroll-target="#format-data">Format data</a></li>
  <li><a href="#split-test-train" id="toc-split-test-train" class="nav-link" data-scroll-target="#split-test-train">Split test &amp; train</a></li>
  <li><a href="#recipe" id="toc-recipe" class="nav-link" data-scroll-target="#recipe">Recipe</a></li>
  <li><a href="#hyper-parameters" id="toc-hyper-parameters" class="nav-link" data-scroll-target="#hyper-parameters">Hyper-parameters</a></li>
  <li><a href="#workflow" id="toc-workflow" class="nav-link" data-scroll-target="#workflow">Workflow</a></li>
  <li><a href="#train-your-model" id="toc-train-your-model" class="nav-link" data-scroll-target="#train-your-model">Train your model!</a></li>
  <li><a href="#predict-the-test-set" id="toc-predict-the-test-set" class="nav-link" data-scroll-target="#predict-the-test-set">Predict the test set!</a></li>
  <li><a href="#evaluate" id="toc-evaluate" class="nav-link" data-scroll-target="#evaluate">Evaluate</a>
  <ul class="collapse">
  <li><a href="#round-probabilities" id="toc-round-probabilities" class="nav-link" data-scroll-target="#round-probabilities">Round probabilities</a></li>
  <li><a href="#boxplot" id="toc-boxplot" class="nav-link" data-scroll-target="#boxplot">Boxplot</a></li>
  <li><a href="#confusion-matrix" id="toc-confusion-matrix" class="nav-link" data-scroll-target="#confusion-matrix">Confusion matrix</a></li>
  <li><a href="#accuracy" id="toc-accuracy" class="nav-link" data-scroll-target="#accuracy">Accuracy</a></li>
  <li><a href="#curves" id="toc-curves" class="nav-link" data-scroll-target="#curves">Curves</a></li>
  <li><a href="#area-under-the-curve" id="toc-area-under-the-curve" class="nav-link" data-scroll-target="#area-under-the-curve">Area under the curve</a></li>
  <li><a href="#feature-gain" id="toc-feature-gain" class="nav-link" data-scroll-target="#feature-gain">Feature Gain</a></li>
  </ul></li>
  <li><a href="#save-output" id="toc-save-output" class="nav-link" data-scroll-target="#save-output">Save output</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./chapter_8.html">8 Machine learning</a></li><li class="breadcrumb-item"><a href="./ch8_section_2.html">Part 2</a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-mlpart02" class="quarto-section-identifier">Part 2</span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="background" class="level2">
<h2 class="anchored" data-anchor-id="background">Background</h2>
<p>We loaded and explored the histology imaging-derived data in <a href="./ch8_section_1.html">ML Part 1</a>. Through this process we found:</p>
<ul>
<li><p>the data requires log-transformation</p></li>
<li><p>there is little separation in malignant vs benign samples when clustering via principal components analysis</p></li>
<li><p>using linear modeling, several imaging features show significant correlation with tumor status (benign vs malignant)</p></li>
</ul>
<p>Given these findings, we have reason to believe that a machine learning algorithm may be successful in predicting malignant vs benign tumor status.</p>
<p>By contrast, if no differences are seen in PCA or simple testing of linear models, the chance of building a reliable classification model is much lower.</p>
</section>
<section id="aim" class="level2">
<h2 class="anchored" data-anchor-id="aim">Aim</h2>
<p>To train a classification model using the ‘xgboost’ algorithm (extreme gradient boosting), which is a class of algorithms that uses decision trees (similar to ‘random forest’).</p>
<p>Note, our primary aim here is <em>not</em> to understand the individual features that contribute to the benign vs malignant classification, but simply to build a robust classification tool.</p>
</section>
<section id="load-libraries" class="level2">
<h2 class="anchored" data-anchor-id="load-libraries">Load libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidymodels)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(xgboost)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load data</h2>
<p>Preprocessed data from <a href="./ch8_section_1.html">ML Part 1</a> is stored in <code>data_processed/</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>dat_wide_log <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">'data_processed/tumor_data_log_wide.xlsx'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="set-seed" class="level2">
<h2 class="anchored" data-anchor-id="set-seed">Set seed</h2>
<p>‘Reproducible randomness’. This ensures everyone using this tutorial gets the same results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="format-data" class="level2">
<h2 class="anchored" data-anchor-id="format-data">Format data</h2>
<p>Training the ML model requires the outcome (malignant status) encoded as a factor.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Data types and file types">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Data types and file types
</div>
</div>
<div class="callout-body-container callout-body">
<p>Data types aren’t preserved in Excel, csv or tsv format files. They are however preserved in .Rdata and .Rds files.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>dat_wide_log <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) <span class="sc">%&gt;%</span>  <span class="fu">str</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [300 × 5] (S3: tbl_df/tbl/data.frame)
 $ donorid     : chr [1:300] "d001" "d002" "d003" "d004" ...
 $ status      : chr [1:300] "malignant" "malignant" "malignant" "benign" ...
 $ cell_size   : num [1:300] -0.769 0.469 0.68 0.669 1.113 ...
 $ cell_density: num [1:300] -0.1844 0.1517 0.6682 0.2399 -0.0523 ...
 $ nuclear_area: num [1:300] -0.0568 -0.1693 -0.2268 -0.3735 -0.3104 ...</code></pre>
</div>
</div>
<p>Re-establish status as a factor:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>dat_wide_log <span class="ot">&lt;-</span> dat_wide_log <span class="sc">%&gt;%</span> </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">status=</span><span class="fu">factor</span>(status))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Drop donorid from the training data. It encodes no useful infromation for training a model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>data_final <span class="ot">&lt;-</span> dat_wide_log <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(donorid))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="split-test-train" class="level2">
<h2 class="anchored" data-anchor-id="split-test-train">Split test &amp; train</h2>
<p>The default is to split the data into training (75%) and test/hold-out (25%). We stratify by status, to ensure roughly equal numbers of benign and malignant samples in the test and training data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>split_cl <span class="ot">&lt;-</span> <span class="fu">initial_split</span>(data_final, <span class="at">strata =</span> status)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Extract the test and training data to separate objects</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>train_cl <span class="ot">&lt;-</span> <span class="fu">training</span>(split_cl)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>test_cl  <span class="ot">&lt;-</span> <span class="fu">testing</span>(split_cl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check the balance of the status classes (malignant vs benign), remembering that ROC curves are more reliable when classes are balanced.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>train_cl <span class="sc">%&gt;%</span> <span class="fu">count</span>(status)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  status        n
  &lt;fct&gt;     &lt;int&gt;
1 benign      104
2 malignant   120</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>test_cl  <span class="sc">%&gt;%</span> <span class="fu">count</span>(status)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  status        n
  &lt;fct&gt;     &lt;int&gt;
1 benign       35
2 malignant    41</code></pre>
</div>
</div>
<p>This gives reasonably even splits.</p>
</section>
<section id="recipe" class="level2">
<h2 class="anchored" data-anchor-id="recipe">Recipe</h2>
<p>The data recipe stores the model parameters, and the data pre-processing steps. These steps will be performed on any data input to the ML training process.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>rec_cl <span class="ot">&lt;-</span> <span class="fu">recipe</span>(status <span class="sc">~</span> .  , <span class="at">data =</span> train_cl ) <span class="sc">%&gt;%</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_zv</span>(<span class="fu">all_numeric_predictors</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_normalize</span>(<span class="fu">all_numeric_predictors</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_dummy</span>(<span class="fu">all_nominal_predictors</span>()) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The model algebra indicates we want to predict the <code>status</code> column (aka ‘y’ / ‘outcome’) using all available columns (denoted by <code>.</code> ). The data argument <code>train_cl</code> is provided to define the structure of the data (i.e., the colnames for all the features we have available).</p>
<p>The tidymodels package has a number of <code>step_()</code> functions that allow us to chain data transformation instructions together using the pipe <code>%&gt;%</code>, similar to a dplyr ‘chain’.</p>
<p><code>step_zv()</code> - step_zero_variance removes any features that have 0 variance, and therefore encode no information that can help to predict the outcome (which is ‘status’, in this case).</p>
<p><code>step_normalize()</code> performs a scaling normalization. First it centres the data each feature by subtracting the mean from every value, and then transforms the centred data into z scores (i.e., dividing each value by the standard deviation for that feature).</p>
<p><code>step_dummy()</code> - converts nominal (character) data such as sex, into ‘dummy’ numeric variables, to allow their use in the ML training data, which is strictly numeric.</p>
<p>Note we have already log-transformed the data in our exploratory data analysis steps in <a href="./ch8_section_1.html">ML Part 1</a>. However, if we had not done this, we could use <code>step_log()</code> before <code>step_normalize()</code>.</p>
</section>
<section id="hyper-parameters" class="level2">
<h2 class="anchored" data-anchor-id="hyper-parameters">Hyper-parameters</h2>
<p>Hyper-parameters are pre-determined settings that govern how the model learns, and the model ‘architecture’ - akin to the depth and breadth of parameters that can be modified during training. Different model classes have different types of hyper-parameters. In this case for the xgboost algorithm, we pre-determine the number of decision trees that are ‘grown’ during the training steps, the depth ( = maximum number of branch-points) of those trees, and the ‘learning rate’ - which governs the magnitude of updates to the model during training.</p>
<p>We also determine the machine learning engine (‘xgboost’) and the mode (‘classification’) as opposed to ‘regression’.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>xgb_params <span class="ot">&lt;-</span> <span class="fu">boost_tree</span>(<span class="at">trees =</span> <span class="dv">500</span>, </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">tree_depth =</span> <span class="dv">20</span>, </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">learn_rate =</span> <span class="fl">0.01</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_engine</span>(<span class="st">"xgboost"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_mode</span>(<span class="st">"classification"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>One of the great features about tidymodels is the simplicity of using different model classes. All that is required is to edit the <code>set_engine()</code> command to your model of choice! The pre-processing and evaluation steps remain unchanged, and this removes the need to learn a separate R library for each ML engine!</p>
</section>
<section id="workflow" class="level2">
<h2 class="anchored" data-anchor-id="workflow">Workflow</h2>
<p>Now we package the data pre-processing recipe and the hyper-parameters into a workflow, ready for training. This workflow construct ensures that all steps - data pre-processing, model specification and tuning, are pre-defined in a single object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>wf_xgb <span class="ot">&lt;-</span> <span class="fu">workflow</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># model outcome and features &amp; pre-processing steps:</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_recipe</span>(rec_cl) <span class="sc">%&gt;%</span>  </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#hyper-parameters, machine learning model type, and mode:</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_model</span>(xgb_params) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Hyper-parameter tuning">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Hyper-parameter tuning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Even though we are not tuning hyper-parameters in this simple workflow, the tuning process can be captured in the workflow object, as we will see later.</p>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Models within models?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Models within models?
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this tutorial ‘models’ appear in two contexts</p>
<ol type="1">
<li><p>The model algebra (similar to a linear model equation), that specifies what our outcome is (y), and the predictor features we want to use. This is specified in the <code>recipe()</code> step.</p></li>
<li><p>The machine learning model (in this case xgboost), which is specified in the workflow using <code>add_model()</code>.</p></li>
</ol>
</div>
</div>
</section>
<section id="train-your-model" class="level2">
<h2 class="anchored" data-anchor-id="train-your-model">Train your model!</h2>
<p>We provide the workflow and the input data. <code>fit()</code> performs the model training steps. This should take ~30-60 seconds on a standard laptop.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>fit_cl <span class="ot">&lt;-</span> <span class="fu">fit</span>(wf_xgb, <span class="at">data =</span> train_cl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="predict-the-test-set" class="level2">
<h2 class="anchored" data-anchor-id="predict-the-test-set">Predict the test set!</h2>
<p>Apply the newly trained model to predict the malignant status in the original <code>train_cl</code> training data, and the hold-out <code>test_cl</code> dataset. Here we set the type argument to ‘prob’, to generate <em>probabilities</em> of each class, rather than discrete labels (0 or 1).</p>
<p>These probabilities will be used directly for plotting ROC and PR curves (below), and rounded for other diagnostic values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pred_train <span class="ot">&lt;-</span> <span class="fu">predict</span>(fit_cl, train_cl, <span class="at">type =</span> <span class="st">'prob'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(train_cl)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>pred_test <span class="ot">&lt;-</span> <span class="fu">predict</span>(fit_cl, test_cl, <span class="st">'prob'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(test_cl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="evaluate" class="level2">
<h2 class="anchored" data-anchor-id="evaluate">Evaluate</h2>
<section id="round-probabilities" class="level3">
<h3 class="anchored" data-anchor-id="round-probabilities">Round probabilities</h3>
<p>First a column of predicted class (discrete) values are generated by rounding the predicted probability of the malignant class</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pred_train <span class="ot">&lt;-</span> pred_train <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">if_else</span>(<span class="fu">round</span>(.pred_malignant)<span class="sc">==</span><span class="dv">1</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'malignant'</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'benign'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(.pred_class, <span class="at">.before =</span> status) <span class="sc">%&gt;%</span> </span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">factor</span>(.pred_class))</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>pred_test <span class="ot">&lt;-</span> pred_test <span class="sc">%&gt;%</span> </span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">if_else</span>(<span class="fu">round</span>(.pred_malignant)<span class="sc">==</span><span class="dv">1</span>,</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'malignant'</span>,</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'benign'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(.pred_class, <span class="at">.before =</span> status) <span class="sc">%&gt;%</span> </span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">factor</span>(.pred_class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="The rounding decision threshold">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The rounding decision threshold
</div>
</div>
<div class="callout-body-container callout-body">
<p>Creating PR and ROC curves requires varying the probability threshold at which a sample is labelled as TRUE (malignant) or FALSE (benign). What is the default decision threshold for the <code>round()</code> function?</p>
</div>
</div>
</section>
<section id="boxplot" class="level3">
<h3 class="anchored" data-anchor-id="boxplot">Boxplot</h3>
<p>Compare the predicted probability of malignancy to the true sample labels</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pred_train <span class="sc">%&gt;%</span> </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>.pred_malignant, <span class="at">x=</span>status)) <span class="sc">+</span> </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> <span class="fu">geom_jitter</span>(<span class="at">width=</span><span class="fl">0.2</span>, <span class="at">height=</span><span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Predicted probability of malignancy'</span>) <span class="sc">+</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'True sample label'</span>) <span class="sc">+</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Training data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ch8_section_2_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>pred_test <span class="sc">%&gt;%</span> </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>.pred_malignant, <span class="at">x=</span>status)) <span class="sc">+</span> </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> <span class="fu">geom_jitter</span>(<span class="at">width=</span><span class="fl">0.2</span>, <span class="at">height=</span><span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Predicted probability of malignancy'</span>) <span class="sc">+</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'True sample label'</span>) <span class="sc">+</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Test data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ch8_section_2_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="confusion-matrix" class="level3">
<h3 class="anchored" data-anchor-id="confusion-matrix">Confusion matrix</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>pred_train <span class="sc">%&gt;%</span> <span class="fu">conf_mat</span>(<span class="at">truth =</span> status, </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">estimate =</span> .pred_class, </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dnn =</span> <span class="fu">c</span>(<span class="st">'Predicted'</span>,<span class="st">'Truth (TRAINING Data)'</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           Truth (TRAINING Data)
Predicted   benign malignant
  benign       104         0
  malignant      0       120</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>pred_test  <span class="sc">%&gt;%</span> <span class="fu">conf_mat</span>(<span class="at">truth=</span>status, </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">estimate =</span> .pred_class, </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dnn =</span> <span class="fu">c</span>(<span class="st">'Predicted'</span>,<span class="st">'Truth (TEST Data)'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           Truth (TEST Data)
Predicted   benign malignant
  benign        21         6
  malignant     14        35</code></pre>
</div>
</div>
<p>What do you notice about the performance of the model on the training, vs the test data?</p>
</section>
<section id="accuracy" class="level3">
<h3 class="anchored" data-anchor-id="accuracy">Accuracy</h3>
<p>The accuracy is the sum of the correct predictions divided by the total number of samples, where 1 = perfect accuracy.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>pred_train <span class="sc">%&gt;%</span> <span class="fu">metrics</span>(<span class="at">truth =</span> status, <span class="at">estimate =</span> .pred_class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  .metric  .estimator .estimate
  &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt;
1 accuracy binary             1
2 kap      binary             1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>pred_test  <span class="sc">%&gt;%</span> <span class="fu">metrics</span>(<span class="at">truth =</span> status, <span class="at">estimate =</span> .pred_class)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  .metric  .estimator .estimate
  &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt;
1 accuracy binary         0.737
2 kap      binary         0.461</code></pre>
</div>
</div>
</section>
<section id="curves" class="level3">
<h3 class="anchored" data-anchor-id="curves">Curves</h3>
<p>Creating ROC and PR curves require the predicted class probabilities rather than discrete labels.</p>
<p>The <code>roc_curve()</code> and <code>pr_curve()</code> functions from the <code>yardstick</code> package (part of the tidymodels stable) are very handy for calculating the true-positive and false-positive rates as the decision threshold decreases.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>roc_tbl <span class="ot">&lt;-</span> <span class="fu">roc_curve</span>(pred_test, <span class="at">truth =</span> status, .pred_malignant, </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">event_level =</span> <span class="st">'second'</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>pr_tbl <span class="ot">&lt;-</span> <span class="fu">pr_curve</span>(pred_test, <span class="at">truth =</span> status, .pred_malignant, </span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">event_level =</span> <span class="st">'second'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For a quick look at performance, ROC and PR curves can be plotted using the <code>autoplot()</code> function (ggplot2 geoms can be added on for customization):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>roc_tbl <span class="sc">%&gt;%</span> <span class="fu">autoplot</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ch8_section_2_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>pr_tbl <span class="sc">%&gt;%</span> <span class="fu">autoplot</span>() <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.5</span>,<span class="at">lty=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ch8_section_2_files/figure-html/unnamed-chunk-20-2.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="area-under-the-curve" class="level3">
<h3 class="anchored" data-anchor-id="area-under-the-curve">Area under the curve</h3>
<p><code>yardstick</code> also contains <code>roc_auc()</code> and <code>pr_auc</code> to calculate the area under each curve type. Note that the ‘event_level’ argument is the category that we consider ‘TRUE’ (in this case ‘malignant’, which is the second level in the status factor).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># AUROCC</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">roc_auc</span>(<span class="at">data =</span> pred_test, <span class="at">truth =</span> status,  .pred_malignant, <span class="at">event_level =</span> <span class="st">'second'</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  .metric .estimator .estimate
  &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;
1 roc_auc binary         0.804</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># AUPRC</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pr_auc</span>( <span class="at">data =</span> pred_test, <span class="at">truth =</span> status, .pred_malignant, <span class="at">event_level =</span> <span class="st">'second'</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  .metric .estimator .estimate
  &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;
1 pr_auc  binary         0.827</code></pre>
</div>
</div>
</section>
<section id="feature-gain" class="level3">
<h3 class="anchored" data-anchor-id="feature-gain">Feature Gain</h3>
<p>‘Gain’ is a measure of the contribution of each feature to the accuracy of an xgboost model. Understanding the relative contribution of each feature is helpful if we want to create a smaller / lighter model using only the most important predictors, for example. This is also an important aspect of ‘explainable AI’ (xAI) /‘interpretable machine learning’. The <code>vip</code> package (‘variable importance plot’) has a <code>vip()</code> function for creating the characteristic horizontal bar charts:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract fitted xgboost model</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>xgb_fit <span class="ot">&lt;-</span> <span class="fu">extract_fit_parsnip</span>(fit_cl)<span class="sc">$</span>fit</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot Gain  ('variable importance plot' - vip)</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>vip<span class="sc">::</span><span class="fu">vip</span>(xgb_fit, <span class="at">num_features =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ch8_section_2_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>What do you notice about the most important features in this model, when compared to our original data exploration work in <a href="./ch8_section_1.html">ML Part 1</a> using <code>lm_test()</code>?</p>
</section>
</section>
<section id="save-output" class="level2">
<h2 class="anchored" data-anchor-id="save-output">Save output</h2>
<p>Let’s save the recipe <code>rec_cl</code>, and <code>pred_test</code> output from the model, to compare with a ‘tuned’ model in the next step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#training &amp; testing data:</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  train_cl, test_cl, </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>     <span class="co">#recipe object:</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>     rec_cl, </span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>     <span class="co">#predictions on test data:</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>     pred_test, <span class="at">file=</span><span class="st">'data_processed/ml_pt2_objects.Rda'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./ch8_section_1.html" class="pagination-link" aria-label="Part 1">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Part 1</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./ch8_section_3.html" class="pagination-link" aria-label="Part 3">
        <span class="nav-page-text">Part 3</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb39" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this section:"</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Part 2 {.unnumbered #sec-mlpart02}</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a><span class="fu">## Background </span></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>We loaded and explored the histology imaging-derived data in <span class="co">[</span><span class="ot">ML Part 1</span><span class="co">](ch8_section_1.qmd)</span>. Through this process we found:</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>the data requires log-transformation</span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>there is little separation in malignant vs benign samples when clustering via principal components analysis</span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>using linear modeling, several imaging features show significant correlation with tumor status (benign vs malignant)</span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>Given these findings, we have reason to believe that a machine learning algorithm may be successful in predicting malignant vs benign tumor status.</span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a>By contrast, if no differences are seen in PCA or simple testing of linear models, the chance of building a reliable classification model is much lower.</span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a><span class="fu">## Aim</span></span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a>To train a classification model using the 'xgboost' algorithm (extreme gradient boosting), which is a class of algorithms that uses decision trees (similar to 'random forest').</span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a>Note, our primary aim here is *not* to understand the individual features that contribute to the benign vs malignant classification, but simply to build a robust classification tool.</span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a><span class="fu">## Load libraries</span></span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidymodels)</span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(xgboost)</span>
<span id="cb39-44"><a href="#cb39-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-45"><a href="#cb39-45" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span>
<span id="cb39-46"><a href="#cb39-46" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-47"><a href="#cb39-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-48"><a href="#cb39-48" aria-hidden="true" tabindex="-1"></a><span class="fu">## Load data</span></span>
<span id="cb39-49"><a href="#cb39-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-50"><a href="#cb39-50" aria-hidden="true" tabindex="-1"></a>Preprocessed data from <span class="co">[</span><span class="ot">ML Part 1</span><span class="co">](ch8_section_1.qmd)</span> is stored in <span class="in">`data_processed/`</span></span>
<span id="cb39-53"><a href="#cb39-53" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-54"><a href="#cb39-54" aria-hidden="true" tabindex="-1"></a>dat_wide_log <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">'data_processed/tumor_data_log_wide.xlsx'</span>)</span>
<span id="cb39-55"><a href="#cb39-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-56"><a href="#cb39-56" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-57"><a href="#cb39-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-58"><a href="#cb39-58" aria-hidden="true" tabindex="-1"></a><span class="fu">## Set seed</span></span>
<span id="cb39-59"><a href="#cb39-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-60"><a href="#cb39-60" aria-hidden="true" tabindex="-1"></a>'Reproducible randomness'. This ensures everyone using this tutorial gets the same results.</span>
<span id="cb39-61"><a href="#cb39-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-64"><a href="#cb39-64" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-65"><a href="#cb39-65" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb39-66"><a href="#cb39-66" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-67"><a href="#cb39-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-68"><a href="#cb39-68" aria-hidden="true" tabindex="-1"></a><span class="fu">## Format data</span></span>
<span id="cb39-69"><a href="#cb39-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-70"><a href="#cb39-70" aria-hidden="true" tabindex="-1"></a>Training the ML model requires the outcome (malignant status) encoded as a factor. </span>
<span id="cb39-71"><a href="#cb39-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-72"><a href="#cb39-72" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Data types and file types"}</span>
<span id="cb39-73"><a href="#cb39-73" aria-hidden="true" tabindex="-1"></a>Data types aren't preserved in Excel, csv or tsv format files. They are however preserved in .Rdata and .Rds files.</span>
<span id="cb39-74"><a href="#cb39-74" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb39-75"><a href="#cb39-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-78"><a href="#cb39-78" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-79"><a href="#cb39-79" aria-hidden="true" tabindex="-1"></a>dat_wide_log <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) <span class="sc">%&gt;%</span>  <span class="fu">str</span>()</span>
<span id="cb39-80"><a href="#cb39-80" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-81"><a href="#cb39-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-82"><a href="#cb39-82" aria-hidden="true" tabindex="-1"></a>Re-establish status as a factor:</span>
<span id="cb39-85"><a href="#cb39-85" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-86"><a href="#cb39-86" aria-hidden="true" tabindex="-1"></a>dat_wide_log <span class="ot">&lt;-</span> dat_wide_log <span class="sc">%&gt;%</span> </span>
<span id="cb39-87"><a href="#cb39-87" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">status=</span><span class="fu">factor</span>(status))</span>
<span id="cb39-88"><a href="#cb39-88" aria-hidden="true" tabindex="-1"></a>                         </span>
<span id="cb39-89"><a href="#cb39-89" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-90"><a href="#cb39-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-91"><a href="#cb39-91" aria-hidden="true" tabindex="-1"></a>Drop donorid from the training data. It encodes no useful infromation for training a model.</span>
<span id="cb39-92"><a href="#cb39-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-95"><a href="#cb39-95" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-96"><a href="#cb39-96" aria-hidden="true" tabindex="-1"></a>data_final <span class="ot">&lt;-</span> dat_wide_log <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(donorid))</span>
<span id="cb39-97"><a href="#cb39-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-98"><a href="#cb39-98" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-99"><a href="#cb39-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-100"><a href="#cb39-100" aria-hidden="true" tabindex="-1"></a><span class="fu">## Split test &amp; train</span></span>
<span id="cb39-101"><a href="#cb39-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-102"><a href="#cb39-102" aria-hidden="true" tabindex="-1"></a>The default is to split the data into training (75%) and test/hold-out (25%). We stratify by status, to ensure roughly equal numbers of benign and malignant samples in the test and training data</span>
<span id="cb39-105"><a href="#cb39-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-106"><a href="#cb39-106" aria-hidden="true" tabindex="-1"></a>split_cl <span class="ot">&lt;-</span> <span class="fu">initial_split</span>(data_final, <span class="at">strata =</span> status)</span>
<span id="cb39-107"><a href="#cb39-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-108"><a href="#cb39-108" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-109"><a href="#cb39-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-110"><a href="#cb39-110" aria-hidden="true" tabindex="-1"></a>Extract the test and training data to separate objects</span>
<span id="cb39-113"><a href="#cb39-113" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-114"><a href="#cb39-114" aria-hidden="true" tabindex="-1"></a>train_cl <span class="ot">&lt;-</span> <span class="fu">training</span>(split_cl)</span>
<span id="cb39-115"><a href="#cb39-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-116"><a href="#cb39-116" aria-hidden="true" tabindex="-1"></a>test_cl  <span class="ot">&lt;-</span> <span class="fu">testing</span>(split_cl)</span>
<span id="cb39-117"><a href="#cb39-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-118"><a href="#cb39-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-119"><a href="#cb39-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-120"><a href="#cb39-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-121"><a href="#cb39-121" aria-hidden="true" tabindex="-1"></a>Check the balance of the status classes (malignant vs benign), remembering that ROC curves are more reliable when classes are balanced.</span>
<span id="cb39-122"><a href="#cb39-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-125"><a href="#cb39-125" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-126"><a href="#cb39-126" aria-hidden="true" tabindex="-1"></a>train_cl <span class="sc">%&gt;%</span> <span class="fu">count</span>(status)</span>
<span id="cb39-127"><a href="#cb39-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-128"><a href="#cb39-128" aria-hidden="true" tabindex="-1"></a>test_cl  <span class="sc">%&gt;%</span> <span class="fu">count</span>(status)</span>
<span id="cb39-129"><a href="#cb39-129" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-130"><a href="#cb39-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-131"><a href="#cb39-131" aria-hidden="true" tabindex="-1"></a>This gives reasonably even splits.</span>
<span id="cb39-132"><a href="#cb39-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-133"><a href="#cb39-133" aria-hidden="true" tabindex="-1"></a><span class="fu">## Recipe</span></span>
<span id="cb39-134"><a href="#cb39-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-135"><a href="#cb39-135" aria-hidden="true" tabindex="-1"></a>The data recipe stores the model parameters, and the data pre-processing steps. These steps will be performed on any data input to the ML training process.</span>
<span id="cb39-136"><a href="#cb39-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-139"><a href="#cb39-139" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-140"><a href="#cb39-140" aria-hidden="true" tabindex="-1"></a>rec_cl <span class="ot">&lt;-</span> <span class="fu">recipe</span>(status <span class="sc">~</span> .  , <span class="at">data =</span> train_cl ) <span class="sc">%&gt;%</span> </span>
<span id="cb39-141"><a href="#cb39-141" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_zv</span>(<span class="fu">all_numeric_predictors</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb39-142"><a href="#cb39-142" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_normalize</span>(<span class="fu">all_numeric_predictors</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb39-143"><a href="#cb39-143" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_dummy</span>(<span class="fu">all_nominal_predictors</span>()) </span>
<span id="cb39-144"><a href="#cb39-144" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-145"><a href="#cb39-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-146"><a href="#cb39-146" aria-hidden="true" tabindex="-1"></a>The model algebra indicates we want to predict the <span class="in">`status`</span> column (aka 'y' / 'outcome') using all available columns (denoted by <span class="in">`.`</span> ). The data argument <span class="in">`train_cl`</span> is provided to define the structure of the data (i.e., the colnames for all the features we have available).</span>
<span id="cb39-147"><a href="#cb39-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-148"><a href="#cb39-148" aria-hidden="true" tabindex="-1"></a>The tidymodels package has a number of <span class="in">`step_()`</span> functions that allow us to chain data transformation instructions together using the pipe <span class="in">`%&gt;%`</span>, similar to a dplyr 'chain'.</span>
<span id="cb39-149"><a href="#cb39-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-150"><a href="#cb39-150" aria-hidden="true" tabindex="-1"></a><span class="in">`step_zv()`</span> - step_zero_variance removes any features that have 0 variance, and therefore encode no information that can help to predict the outcome (which is 'status', in this case).</span>
<span id="cb39-151"><a href="#cb39-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-152"><a href="#cb39-152" aria-hidden="true" tabindex="-1"></a><span class="in">`step_normalize()`</span> performs a scaling normalization. First it centres the data each feature by subtracting the mean from every value, and then transforms the centred data into z scores (i.e., dividing each value by the standard deviation for that feature).</span>
<span id="cb39-153"><a href="#cb39-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-154"><a href="#cb39-154" aria-hidden="true" tabindex="-1"></a><span class="in">`step_dummy()`</span> - converts nominal (character) data such as sex, into 'dummy' numeric variables, to allow their use in the ML training data, which is strictly numeric.</span>
<span id="cb39-155"><a href="#cb39-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-156"><a href="#cb39-156" aria-hidden="true" tabindex="-1"></a>Note we have already log-transformed the data in our exploratory data analysis steps in <span class="co">[</span><span class="ot">ML Part 1</span><span class="co">](ch8_section_1.qmd)</span>. However, if we had not done this, we could use <span class="in">`step_log()`</span> before <span class="in">`step_normalize()`</span>.</span>
<span id="cb39-157"><a href="#cb39-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-158"><a href="#cb39-158" aria-hidden="true" tabindex="-1"></a><span class="fu">## Hyper-parameters</span></span>
<span id="cb39-159"><a href="#cb39-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-160"><a href="#cb39-160" aria-hidden="true" tabindex="-1"></a>Hyper-parameters are pre-determined settings that govern how the model learns, and the model 'architecture' - akin to the depth and breadth of parameters that can be modified during training. </span>
<span id="cb39-161"><a href="#cb39-161" aria-hidden="true" tabindex="-1"></a>Different model classes have different types of hyper-parameters. In this case for the xgboost algorithm, we pre-determine the number of decision trees that are 'grown' during the training steps, the depth ( = maximum number of branch-points) of those trees, and the 'learning rate' - which governs the magnitude of updates to the model during training.</span>
<span id="cb39-162"><a href="#cb39-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-163"><a href="#cb39-163" aria-hidden="true" tabindex="-1"></a>We also determine the machine learning engine ('xgboost') and the mode ('classification') as opposed to 'regression'.</span>
<span id="cb39-164"><a href="#cb39-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-167"><a href="#cb39-167" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-168"><a href="#cb39-168" aria-hidden="true" tabindex="-1"></a>xgb_params <span class="ot">&lt;-</span> <span class="fu">boost_tree</span>(<span class="at">trees =</span> <span class="dv">500</span>, </span>
<span id="cb39-169"><a href="#cb39-169" aria-hidden="true" tabindex="-1"></a>                           <span class="at">tree_depth =</span> <span class="dv">20</span>, </span>
<span id="cb39-170"><a href="#cb39-170" aria-hidden="true" tabindex="-1"></a>                          <span class="at">learn_rate =</span> <span class="fl">0.01</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-171"><a href="#cb39-171" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_engine</span>(<span class="st">"xgboost"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-172"><a href="#cb39-172" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_mode</span>(<span class="st">"classification"</span>)</span>
<span id="cb39-173"><a href="#cb39-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-174"><a href="#cb39-174" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-175"><a href="#cb39-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-176"><a href="#cb39-176" aria-hidden="true" tabindex="-1"></a>One of the great features about tidymodels is the simplicity of using different model classes. All that is required is to edit the <span class="in">`set_engine()`</span> command to your model of choice! The pre-processing and evaluation steps remain unchanged, and this removes the need to learn a separate R library for each ML engine!</span>
<span id="cb39-177"><a href="#cb39-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-178"><a href="#cb39-178" aria-hidden="true" tabindex="-1"></a><span class="fu">## Workflow</span></span>
<span id="cb39-179"><a href="#cb39-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-180"><a href="#cb39-180" aria-hidden="true" tabindex="-1"></a>Now we package the data pre-processing recipe and the hyper-parameters into a workflow, ready for training. This workflow construct ensures that all steps - data pre-processing, model specification and tuning, are pre-defined in a single object.</span>
<span id="cb39-181"><a href="#cb39-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-184"><a href="#cb39-184" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-185"><a href="#cb39-185" aria-hidden="true" tabindex="-1"></a>wf_xgb <span class="ot">&lt;-</span> <span class="fu">workflow</span>() <span class="sc">%&gt;%</span></span>
<span id="cb39-186"><a href="#cb39-186" aria-hidden="true" tabindex="-1"></a>  <span class="co"># model outcome and features &amp; pre-processing steps:</span></span>
<span id="cb39-187"><a href="#cb39-187" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_recipe</span>(rec_cl) <span class="sc">%&gt;%</span>  </span>
<span id="cb39-188"><a href="#cb39-188" aria-hidden="true" tabindex="-1"></a>  <span class="co">#hyper-parameters, machine learning model type, and mode:</span></span>
<span id="cb39-189"><a href="#cb39-189" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_model</span>(xgb_params) </span>
<span id="cb39-190"><a href="#cb39-190" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-191"><a href="#cb39-191" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-192"><a href="#cb39-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-193"><a href="#cb39-193" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Hyper-parameter tuning"}</span>
<span id="cb39-194"><a href="#cb39-194" aria-hidden="true" tabindex="-1"></a>Even though we are not tuning hyper-parameters in this simple workflow, the tuning process can be captured in the workflow object, as we will see later.</span>
<span id="cb39-195"><a href="#cb39-195" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb39-196"><a href="#cb39-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-197"><a href="#cb39-197" aria-hidden="true" tabindex="-1"></a>::: {.callout-important title="Models within models?"}</span>
<span id="cb39-198"><a href="#cb39-198" aria-hidden="true" tabindex="-1"></a>In this tutorial 'models' appear in two contexts</span>
<span id="cb39-199"><a href="#cb39-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-200"><a href="#cb39-200" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The model algebra (similar to a linear model equation), that specifies what our outcome is (y), and the predictor features we want to use. This is specified in the <span class="in">`recipe()`</span> step.</span>
<span id="cb39-201"><a href="#cb39-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-202"><a href="#cb39-202" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The machine learning model (in this case xgboost), which is specified in the workflow using <span class="in">`add_model()`</span>.</span>
<span id="cb39-203"><a href="#cb39-203" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb39-204"><a href="#cb39-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-205"><a href="#cb39-205" aria-hidden="true" tabindex="-1"></a><span class="fu">## Train your model!</span></span>
<span id="cb39-206"><a href="#cb39-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-207"><a href="#cb39-207" aria-hidden="true" tabindex="-1"></a>We provide the workflow and the input data. <span class="in">`fit()`</span> performs the model training steps. This should take \~30-60 seconds on a standard laptop.</span>
<span id="cb39-208"><a href="#cb39-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-211"><a href="#cb39-211" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-212"><a href="#cb39-212" aria-hidden="true" tabindex="-1"></a>fit_cl <span class="ot">&lt;-</span> <span class="fu">fit</span>(wf_xgb, <span class="at">data =</span> train_cl)</span>
<span id="cb39-213"><a href="#cb39-213" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-214"><a href="#cb39-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-215"><a href="#cb39-215" aria-hidden="true" tabindex="-1"></a><span class="fu">## Predict the test set!</span></span>
<span id="cb39-216"><a href="#cb39-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-217"><a href="#cb39-217" aria-hidden="true" tabindex="-1"></a>Apply the newly trained model to predict the malignant status in the original <span class="in">`train_cl`</span> training data, and the hold-out <span class="in">`test_cl`</span> dataset. Here we set the type argument to 'prob', to generate _probabilities_ of each class, rather than discrete labels (0 or 1).</span>
<span id="cb39-218"><a href="#cb39-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-219"><a href="#cb39-219" aria-hidden="true" tabindex="-1"></a>These probabilities will be used directly for plotting ROC and PR curves (below), and rounded for other diagnostic values.</span>
<span id="cb39-220"><a href="#cb39-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-223"><a href="#cb39-223" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-224"><a href="#cb39-224" aria-hidden="true" tabindex="-1"></a>pred_train <span class="ot">&lt;-</span> <span class="fu">predict</span>(fit_cl, train_cl, <span class="at">type =</span> <span class="st">'prob'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-225"><a href="#cb39-225" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(train_cl)</span>
<span id="cb39-226"><a href="#cb39-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-227"><a href="#cb39-227" aria-hidden="true" tabindex="-1"></a>pred_test <span class="ot">&lt;-</span> <span class="fu">predict</span>(fit_cl, test_cl, <span class="st">'prob'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-228"><a href="#cb39-228" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(test_cl)</span>
<span id="cb39-229"><a href="#cb39-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-230"><a href="#cb39-230" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-231"><a href="#cb39-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-232"><a href="#cb39-232" aria-hidden="true" tabindex="-1"></a><span class="fu">## Evaluate</span></span>
<span id="cb39-233"><a href="#cb39-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-234"><a href="#cb39-234" aria-hidden="true" tabindex="-1"></a><span class="fu">### Round probabilities</span></span>
<span id="cb39-235"><a href="#cb39-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-236"><a href="#cb39-236" aria-hidden="true" tabindex="-1"></a>First a column of predicted class (discrete) values are generated by rounding the predicted probability of the malignant class</span>
<span id="cb39-237"><a href="#cb39-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-240"><a href="#cb39-240" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-241"><a href="#cb39-241" aria-hidden="true" tabindex="-1"></a>pred_train <span class="ot">&lt;-</span> pred_train <span class="sc">%&gt;%</span> </span>
<span id="cb39-242"><a href="#cb39-242" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">if_else</span>(<span class="fu">round</span>(.pred_malignant)<span class="sc">==</span><span class="dv">1</span>,</span>
<span id="cb39-243"><a href="#cb39-243" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'malignant'</span>,</span>
<span id="cb39-244"><a href="#cb39-244" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'benign'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-245"><a href="#cb39-245" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(.pred_class, <span class="at">.before =</span> status) <span class="sc">%&gt;%</span> </span>
<span id="cb39-246"><a href="#cb39-246" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">factor</span>(.pred_class))</span>
<span id="cb39-247"><a href="#cb39-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-248"><a href="#cb39-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-249"><a href="#cb39-249" aria-hidden="true" tabindex="-1"></a>pred_test <span class="ot">&lt;-</span> pred_test <span class="sc">%&gt;%</span> </span>
<span id="cb39-250"><a href="#cb39-250" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">if_else</span>(<span class="fu">round</span>(.pred_malignant)<span class="sc">==</span><span class="dv">1</span>,</span>
<span id="cb39-251"><a href="#cb39-251" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'malignant'</span>,</span>
<span id="cb39-252"><a href="#cb39-252" aria-hidden="true" tabindex="-1"></a>                             <span class="st">'benign'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-253"><a href="#cb39-253" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(.pred_class, <span class="at">.before =</span> status) <span class="sc">%&gt;%</span> </span>
<span id="cb39-254"><a href="#cb39-254" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">.pred_class=</span><span class="fu">factor</span>(.pred_class))</span>
<span id="cb39-255"><a href="#cb39-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-256"><a href="#cb39-256" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-257"><a href="#cb39-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-258"><a href="#cb39-258" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="The rounding decision threshold"}</span>
<span id="cb39-259"><a href="#cb39-259" aria-hidden="true" tabindex="-1"></a>Creating PR and ROC curves requires varying the probability threshold at which a sample is labelled as TRUE (malignant) or FALSE (benign). What is the default decision threshold for the <span class="in">`round()`</span> function?</span>
<span id="cb39-260"><a href="#cb39-260" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb39-261"><a href="#cb39-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-262"><a href="#cb39-262" aria-hidden="true" tabindex="-1"></a><span class="fu">### Boxplot</span></span>
<span id="cb39-263"><a href="#cb39-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-264"><a href="#cb39-264" aria-hidden="true" tabindex="-1"></a>Compare the predicted probability of malignancy to the true sample labels</span>
<span id="cb39-265"><a href="#cb39-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-266"><a href="#cb39-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-269"><a href="#cb39-269" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-270"><a href="#cb39-270" aria-hidden="true" tabindex="-1"></a>pred_train <span class="sc">%&gt;%</span> </span>
<span id="cb39-271"><a href="#cb39-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>.pred_malignant, <span class="at">x=</span>status)) <span class="sc">+</span> </span>
<span id="cb39-272"><a href="#cb39-272" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> <span class="fu">geom_jitter</span>(<span class="at">width=</span><span class="fl">0.2</span>, <span class="at">height=</span><span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb39-273"><a href="#cb39-273" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Predicted probability of malignancy'</span>) <span class="sc">+</span></span>
<span id="cb39-274"><a href="#cb39-274" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'True sample label'</span>) <span class="sc">+</span></span>
<span id="cb39-275"><a href="#cb39-275" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Training data'</span>)</span>
<span id="cb39-276"><a href="#cb39-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-277"><a href="#cb39-277" aria-hidden="true" tabindex="-1"></a>pred_test <span class="sc">%&gt;%</span> </span>
<span id="cb39-278"><a href="#cb39-278" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>.pred_malignant, <span class="at">x=</span>status)) <span class="sc">+</span> </span>
<span id="cb39-279"><a href="#cb39-279" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> <span class="fu">geom_jitter</span>(<span class="at">width=</span><span class="fl">0.2</span>, <span class="at">height=</span><span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb39-280"><a href="#cb39-280" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Predicted probability of malignancy'</span>) <span class="sc">+</span></span>
<span id="cb39-281"><a href="#cb39-281" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'True sample label'</span>) <span class="sc">+</span></span>
<span id="cb39-282"><a href="#cb39-282" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Test data'</span>)</span>
<span id="cb39-283"><a href="#cb39-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-284"><a href="#cb39-284" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-285"><a href="#cb39-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-286"><a href="#cb39-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-287"><a href="#cb39-287" aria-hidden="true" tabindex="-1"></a><span class="fu">### Confusion matrix</span></span>
<span id="cb39-288"><a href="#cb39-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-289"><a href="#cb39-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-292"><a href="#cb39-292" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-293"><a href="#cb39-293" aria-hidden="true" tabindex="-1"></a>pred_train <span class="sc">%&gt;%</span> <span class="fu">conf_mat</span>(<span class="at">truth =</span> status, </span>
<span id="cb39-294"><a href="#cb39-294" aria-hidden="true" tabindex="-1"></a>                        <span class="at">estimate =</span> .pred_class, </span>
<span id="cb39-295"><a href="#cb39-295" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dnn =</span> <span class="fu">c</span>(<span class="st">'Predicted'</span>,<span class="st">'Truth (TRAINING Data)'</span>)) </span>
<span id="cb39-296"><a href="#cb39-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-297"><a href="#cb39-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-298"><a href="#cb39-298" aria-hidden="true" tabindex="-1"></a>pred_test  <span class="sc">%&gt;%</span> <span class="fu">conf_mat</span>(<span class="at">truth=</span>status, </span>
<span id="cb39-299"><a href="#cb39-299" aria-hidden="true" tabindex="-1"></a>                        <span class="at">estimate =</span> .pred_class, </span>
<span id="cb39-300"><a href="#cb39-300" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dnn =</span> <span class="fu">c</span>(<span class="st">'Predicted'</span>,<span class="st">'Truth (TEST Data)'</span>))</span>
<span id="cb39-301"><a href="#cb39-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-302"><a href="#cb39-302" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-303"><a href="#cb39-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-304"><a href="#cb39-304" aria-hidden="true" tabindex="-1"></a>What do you notice about the performance of the model on the training, vs the test data?</span>
<span id="cb39-305"><a href="#cb39-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-306"><a href="#cb39-306" aria-hidden="true" tabindex="-1"></a><span class="fu">### Accuracy</span></span>
<span id="cb39-307"><a href="#cb39-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-308"><a href="#cb39-308" aria-hidden="true" tabindex="-1"></a>The accuracy is the sum of the correct predictions divided by the total number of samples, where 1 = perfect accuracy.</span>
<span id="cb39-309"><a href="#cb39-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-312"><a href="#cb39-312" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-313"><a href="#cb39-313" aria-hidden="true" tabindex="-1"></a>pred_train <span class="sc">%&gt;%</span> <span class="fu">metrics</span>(<span class="at">truth =</span> status, <span class="at">estimate =</span> .pred_class)</span>
<span id="cb39-314"><a href="#cb39-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-315"><a href="#cb39-315" aria-hidden="true" tabindex="-1"></a>pred_test  <span class="sc">%&gt;%</span> <span class="fu">metrics</span>(<span class="at">truth =</span> status, <span class="at">estimate =</span> .pred_class)</span>
<span id="cb39-316"><a href="#cb39-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-317"><a href="#cb39-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-318"><a href="#cb39-318" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-319"><a href="#cb39-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-320"><a href="#cb39-320" aria-hidden="true" tabindex="-1"></a><span class="fu">### Curves</span></span>
<span id="cb39-321"><a href="#cb39-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-322"><a href="#cb39-322" aria-hidden="true" tabindex="-1"></a>Creating ROC and PR curves require the predicted class probabilities rather than discrete labels. </span>
<span id="cb39-323"><a href="#cb39-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-324"><a href="#cb39-324" aria-hidden="true" tabindex="-1"></a>The <span class="in">`roc_curve()`</span> and <span class="in">`pr_curve()`</span> functions from the <span class="in">`yardstick`</span> package (part of the tidymodels stable) are very handy for calculating the true-positive and false-positive rates as the decision threshold decreases.</span>
<span id="cb39-325"><a href="#cb39-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-328"><a href="#cb39-328" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-329"><a href="#cb39-329" aria-hidden="true" tabindex="-1"></a>roc_tbl <span class="ot">&lt;-</span> <span class="fu">roc_curve</span>(pred_test, <span class="at">truth =</span> status, .pred_malignant, </span>
<span id="cb39-330"><a href="#cb39-330" aria-hidden="true" tabindex="-1"></a>                     <span class="at">event_level =</span> <span class="st">'second'</span>)</span>
<span id="cb39-331"><a href="#cb39-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-332"><a href="#cb39-332" aria-hidden="true" tabindex="-1"></a>pr_tbl <span class="ot">&lt;-</span> <span class="fu">pr_curve</span>(pred_test, <span class="at">truth =</span> status, .pred_malignant, </span>
<span id="cb39-333"><a href="#cb39-333" aria-hidden="true" tabindex="-1"></a>                     <span class="at">event_level =</span> <span class="st">'second'</span>)</span>
<span id="cb39-334"><a href="#cb39-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-335"><a href="#cb39-335" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-336"><a href="#cb39-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-337"><a href="#cb39-337" aria-hidden="true" tabindex="-1"></a>For a quick look at performance, ROC and PR curves can be plotted using the <span class="in">`autoplot()`</span> function (ggplot2 geoms can be added on for customization):</span>
<span id="cb39-340"><a href="#cb39-340" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-341"><a href="#cb39-341" aria-hidden="true" tabindex="-1"></a>roc_tbl <span class="sc">%&gt;%</span> <span class="fu">autoplot</span>() </span>
<span id="cb39-342"><a href="#cb39-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-343"><a href="#cb39-343" aria-hidden="true" tabindex="-1"></a>pr_tbl <span class="sc">%&gt;%</span> <span class="fu">autoplot</span>() <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.5</span>,<span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb39-344"><a href="#cb39-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-345"><a href="#cb39-345" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-346"><a href="#cb39-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-347"><a href="#cb39-347" aria-hidden="true" tabindex="-1"></a><span class="fu">### Area under the curve</span></span>
<span id="cb39-348"><a href="#cb39-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-349"><a href="#cb39-349" aria-hidden="true" tabindex="-1"></a><span class="in">`yardstick`</span> also contains <span class="in">`roc_auc()`</span> and <span class="in">`pr_auc`</span> to calculate the area under each curve type. Note that the 'event_level' argument is the category that we consider 'TRUE' (in this case 'malignant', which is the second level in the status factor).</span>
<span id="cb39-352"><a href="#cb39-352" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-353"><a href="#cb39-353" aria-hidden="true" tabindex="-1"></a><span class="co"># AUROCC</span></span>
<span id="cb39-354"><a href="#cb39-354" aria-hidden="true" tabindex="-1"></a><span class="fu">roc_auc</span>(<span class="at">data =</span> pred_test, <span class="at">truth =</span> status,  .pred_malignant, <span class="at">event_level =</span> <span class="st">'second'</span>) </span>
<span id="cb39-355"><a href="#cb39-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-356"><a href="#cb39-356" aria-hidden="true" tabindex="-1"></a><span class="co"># AUPRC</span></span>
<span id="cb39-357"><a href="#cb39-357" aria-hidden="true" tabindex="-1"></a><span class="fu">pr_auc</span>( <span class="at">data =</span> pred_test, <span class="at">truth =</span> status, .pred_malignant, <span class="at">event_level =</span> <span class="st">'second'</span>) </span>
<span id="cb39-358"><a href="#cb39-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-359"><a href="#cb39-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-360"><a href="#cb39-360" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-361"><a href="#cb39-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-362"><a href="#cb39-362" aria-hidden="true" tabindex="-1"></a><span class="fu">### Feature Gain</span></span>
<span id="cb39-363"><a href="#cb39-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-364"><a href="#cb39-364" aria-hidden="true" tabindex="-1"></a>'Gain' is a measure of the contribution of each feature to the accuracy of an xgboost model. </span>
<span id="cb39-365"><a href="#cb39-365" aria-hidden="true" tabindex="-1"></a>Understanding the relative contribution of each feature is helpful if we want to create a smaller / lighter model using only the most important predictors, for example. This is also an important aspect of 'explainable AI' (xAI) /'interpretable machine learning'.</span>
<span id="cb39-366"><a href="#cb39-366" aria-hidden="true" tabindex="-1"></a>The <span class="in">`vip`</span> package ('variable importance plot') has a <span class="in">`vip()`</span> function for creating the characteristic horizontal bar charts: </span>
<span id="cb39-367"><a href="#cb39-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-370"><a href="#cb39-370" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-371"><a href="#cb39-371" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract fitted xgboost model</span></span>
<span id="cb39-372"><a href="#cb39-372" aria-hidden="true" tabindex="-1"></a>xgb_fit <span class="ot">&lt;-</span> <span class="fu">extract_fit_parsnip</span>(fit_cl)<span class="sc">$</span>fit</span>
<span id="cb39-373"><a href="#cb39-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-374"><a href="#cb39-374" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot Gain  ('variable importance plot' - vip)</span></span>
<span id="cb39-375"><a href="#cb39-375" aria-hidden="true" tabindex="-1"></a>vip<span class="sc">::</span><span class="fu">vip</span>(xgb_fit, <span class="at">num_features =</span> <span class="dv">20</span>)</span>
<span id="cb39-376"><a href="#cb39-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-377"><a href="#cb39-377" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb39-378"><a href="#cb39-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-379"><a href="#cb39-379" aria-hidden="true" tabindex="-1"></a>What do you notice about the most important features in this model, when compared to our original data exploration work in <span class="co">[</span><span class="ot">ML Part 1</span><span class="co">](ch8_section_1.qmd)</span> using <span class="in">`lm_test()`</span>?</span>
<span id="cb39-380"><a href="#cb39-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-381"><a href="#cb39-381" aria-hidden="true" tabindex="-1"></a><span class="fu">## Save output</span></span>
<span id="cb39-382"><a href="#cb39-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-383"><a href="#cb39-383" aria-hidden="true" tabindex="-1"></a>Let's save the recipe <span class="in">`rec_cl`</span>, and <span class="in">`pred_test`</span> output from the model, to compare with a 'tuned' model in the next step.</span>
<span id="cb39-384"><a href="#cb39-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-387"><a href="#cb39-387" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb39-388"><a href="#cb39-388" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb39-389"><a href="#cb39-389" aria-hidden="true" tabindex="-1"></a>  <span class="co">#training &amp; testing data:</span></span>
<span id="cb39-390"><a href="#cb39-390" aria-hidden="true" tabindex="-1"></a>  train_cl, test_cl, </span>
<span id="cb39-391"><a href="#cb39-391" aria-hidden="true" tabindex="-1"></a>     <span class="co">#recipe object:</span></span>
<span id="cb39-392"><a href="#cb39-392" aria-hidden="true" tabindex="-1"></a>     rec_cl, </span>
<span id="cb39-393"><a href="#cb39-393" aria-hidden="true" tabindex="-1"></a>     <span class="co">#predictions on test data:</span></span>
<span id="cb39-394"><a href="#cb39-394" aria-hidden="true" tabindex="-1"></a>     pred_test, <span class="at">file=</span><span class="st">'data_processed/ml_pt2_objects.Rda'</span>)</span>
<span id="cb39-395"><a href="#cb39-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-396"><a href="#cb39-396" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>