<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; Working with data - Part 1 – BIOL90042 R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./chapter_3.html" rel="next">
<link href="./chapter_1.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-063905fb60d21bb612e30333561bc9be.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-ee22da2db2cf0c267ae220c203eece4d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<script src="site_libs/quarto-contrib/kbd/kbd.js"></script>
<link href="site_libs/quarto-contrib/kbd/kbd.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./chapter_2.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Working with data - Part 1</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">BIOL90042 R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Working with data - Part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Working with data - Part 2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Plotting with ggplot2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Putting it all together</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./stats_primer.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Statistics in R Primer</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">RNA-seq - Part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">RNA-seq - Part 2</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./chapter_8.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">8 Machine learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 3</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./further_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Further reading</span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this chapter:</h2>
   
  <ul>
  <li><a href="#sec-pipes" id="toc-sec-pipes" class="nav-link active" data-scroll-target="#sec-pipes"><span class="header-section-number">2.1</span> Chaining functions together with pipes</a></li>
  <li><a href="#sec-dataManip" id="toc-sec-dataManip" class="nav-link" data-scroll-target="#sec-dataManip"><span class="header-section-number">2.2</span> Basic data manipulation</a>
  <ul class="collapse">
  <li><a href="#sec-sorting" id="toc-sec-sorting" class="nav-link" data-scroll-target="#sec-sorting"><span class="header-section-number">2.2.1</span> Sorting data</a></li>
  <li><a href="#sec-filter" id="toc-sec-filter" class="nav-link" data-scroll-target="#sec-filter"><span class="header-section-number">2.2.2</span> Filtering data (rows)</a></li>
  <li><a href="#sec-missing" id="toc-sec-missing" class="nav-link" data-scroll-target="#sec-missing"><span class="header-section-number">2.2.3</span> Dealing with missing values</a></li>
  <li><a href="#sec-select" id="toc-sec-select" class="nav-link" data-scroll-target="#sec-select"><span class="header-section-number">2.2.4</span> Selecting columns</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">2.3</span> Summary</a>
  <ul class="collapse">
  <li><a href="#practice-questions" id="toc-practice-questions" class="nav-link" data-scroll-target="#practice-questions"><span class="header-section-number">2.3.1</span> Practice questions</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-chapter02" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Working with data - Part 1</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this chapter we will learn how to manipulate and summarise data using the <code>dplyr</code> package (with a little help from the <code>tidyr</code> package too).</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the end of this chapter, learners should be able to:</p>
<ol type="1">
<li><p>Use the pipe (<code>%&gt;%</code>) to chain multiple functions together</p></li>
<li><p>Design chains of dplyr functions to manipulate data frames</p></li>
<li><p>Understand how to identify and handle missing values in a data frame</p></li>
<li><p>Apply grouping for more complex analysis of data</p></li>
<li><p>Recall how to save data frames to a file</p></li>
</ol>
</div>
</div>
<p>Both <code>dplyr</code> and <code>tidyr</code> are contained within the <code>tidyverse</code> (along with <code>readr</code>) so we can load all of these packages at once using <code>library(tidyverse)</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># don't forget to load tidyverse!</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
</div>
<section id="sec-pipes" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sec-pipes"><span class="header-section-number">2.1</span> Chaining functions together with pipes</h2>
<p>Pipes are a powerful feature of the <code>tidyverse</code> that allow you to chain multiple functions together. Pipes are useful because they allow you to break down complex operations into smaller steps that are easier to read and understand.</p>
<p>For example, take the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(<span class="fu">round</span>(<span class="fu">mean</span>(my_vector)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>What do you think this code does? It calculates the mean of <code>my_vector</code>, rounds the result to the nearest whole number, and then converts the result to a character. But the code is a bit hard to read because you have to start from the inside of the brackets and work your way out.</p>
<p>Instead, we can use the pipe operator (<code>%&gt;%</code>) to chain these functions together in a more readable way:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>() <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>See how the code reads naturally from left to right? You can think of the pipe as being like the phrase “and then”. Here, we’re telling R: “Take <code>my_vector</code>, and then calculate the mean, and then round the result, and then convert it to a character.”</p>
<p>You’ll notice that we didn’t need to specify the input to each function. That’s because the pipe automatically passes the output of the previous function as the first input to the next function. We can still specify additional arguments to each function if we need to. For example, if we wanted to round the mean to 2 decimal places, we could do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3"</code></pre>
</div>
</div>
<p>R is clever enough to know that the first argument to <code>round()</code> is still the output of <code>mean()</code>, even though we’ve now specified the <code>digits</code> argument.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Plenty of pipes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Plenty of pipes
</div>
</div>
<div class="callout-body-container callout-body">
<p>There is another style of pipe in R, called the ‘base R pipe’ <code>|&gt;</code>, which is available in R version 4.1.0 and later. The base R pipe works in a similar way to the <code>magrittr</code> pipe (<code>%&gt;%</code>) that we use in this course, but it is not as flexible. We recommend using the <code>magrittr</code> pipe for now.</p>
<p>Fun fact: the <code>magrittr</code> package is named after the <a href="https://en.wikipedia.org/wiki/The_Treachery_of_Images">artist René Magritte, who made a famous painting of a pipe</a>.</p>
</div>
</div>
<p>To type the pipe operator more easily, you can use the keyboard shortcut <kbd aria-hidden="true">Cmd-shift-M</kbd><span class="visually-hidden">Cmd-shift-M</span> (although once you get used to it, you might find it easier to type <code>%&gt;%</code> manually).</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What is NOT a valid way to re-write the following code using the pipe operator: <code>round(sqrt(sum(1:10)), 1)</code>. If you’re not sure, try running the different options in the console to see which one gives the same answer.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(1)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>sum(1:10) %&gt;% sqrt(1) %&gt;% round()</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(digits = 1)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>sum(1:10) %&gt;% sqrt() %&gt;% round(digits = 1)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is the output of the following code? <code>letters %&gt;% head() %&gt;% toupper()</code> Try to guess it before copy-pasting into R.</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>"A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z"</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>"a" "b" "c" "d" "e" "f"</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>An error</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>"A" "B" "C" "D" "E" "F"</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The invalid option is <code>sum(1:10) %&gt;% sqrt(1) %&gt;% round()</code>. This is because the <code>sqrt()</code> function only takes one argument, so you can’t specify <code>1</code> as an argument in addition to what is being piped in from <code>sum(1:10)</code>. Note that some options used the pipe to send <code>1:10</code> to <code>sum()</code> (like <code>1:10 %&gt;% sum()</code>), and others just used <code>sum(1:10)</code> directly. Both are valid ways to use the pipe, it’s just a matter of personal preference.</li>
<li>The output of the code <code>letters %&gt;% head() %&gt;% toupper()</code> is <code>"A" "B" "C" "D" "E" "F"</code>. The <code>letters</code> vector contains the lowercase alphabet, and the <code>head()</code> function returns the first 6 elements of the vector. Finally, the <code>toupper()</code> function then converts these elements to uppercase.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-dataManip" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="sec-dataManip"><span class="header-section-number">2.2</span> Basic data manipulation</h2>
<p>To really see the power of the pipe, we will use it together with the <code>dplyr</code> package that provides a set of functions to easily filter, sort, select, modify and summarise data frames. These functions are designed to work well with the pipe, so you can chain them together to create complex data manipulations in a readable format.</p>
<p>For example, even though we haven’t covered the <code>dplyr</code> functions yet, you can probably guess what the following code does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use the pipe to chain together our data manipulation steps</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This code filters the <code>m_dose</code> data frame to only include data from cage 3E, then pulls out the <code>weight_lost_g</code> column, and finally calculates the mean of the values in that column. The first argument to each function is the output of the previous function, and any additional arguments (like the column name in <code>pull()</code>) are specified in the brackets (like <code>round(digits = 2)</code> from the previous example).</p>
<p>We also used the enter key after each pipe <code>%&gt;%</code> to break up the code into multiple lines to make it easier to read. This isn’t required, but is a popular style in the R community, so all the code examples in this chapter will follow this format.</p>
<p>We will now introduce some of the most commonly used <code>dplyr</code> functions for manipulating data frames. To showcase these, we will use the <code>m_dose</code> that we practiced reading in last chapter. This imaginary dataset contains information on the weight lost by different strains of mice after being treated with different doses of MouseZempic®.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data, like we did in chapter 1</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 344 Columns: 9
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): mouse_strain, cage_number, replicate, sex
dbl (5): weight_lost_g, drug_dose_g, tail_length_mm, initial_weight_g, id_num

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>Before we start, let’s use what we learned in the previous chapter to take a look at <code>m_dose</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># it's a tibble, so prints nicely</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>m_dose</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>You might also like to use <code>View()</code> to open the data in a separate window and get a closer look.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Using RStudio autocomplete">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Using RStudio autocomplete
</div>
</div>
<div class="callout-body-container callout-body">
<p>Although it’s great to give our data a descriptive name like <code>m_dose</code>, it can be a bit of a pain to type out every time. Luckily, RStudio has a handy autocomplete feature that can solve this problem. Just start typing the name of the object, and you’ll see it will popup:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F01_rstudio_autocomplete.png" class="img-fluid figure-img"></p>
<figcaption>RStudio autocomplete</figcaption>
</figure>
</div>
<p>You can then press <kbd aria-hidden="true">Tab</kbd><span class="visually-hidden">Tab</span> to autocomplete it. If there are multiple objects that start with the same letters, you can use the arrow keys to cycle through the options.</p>
<p>Try using autocomplete in this chapter to save yourself some typing!</p>
</div>
</div>
<section id="sec-sorting" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="sec-sorting"><span class="header-section-number">2.2.1</span> Sorting data</h3>
<p>Often, one of the first things you might want to do with a dataset is sort it. In <code>dplyr</code>, this is called ‘arranging’ and is done with the <code>arrange()</code> function.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F02_arrange.png" class="img-fluid figure-img"></p>
<figcaption>Arrange orders rows by their values in one or more columns</figcaption>
</figure>
</div>
<p>By default, <code>arrange()</code> sorts in ascending order (smallest values first). For example, let’s sort the <code>m_dose</code> data frame by the <code>weight_lost_g</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 BALB C       2B                   2.7  rep2      female     0.00192
 2 CD-1         3E                   2.85 rep2      female     0.00181
 3 CD-1         3E                   2.85 rep2      female     0.00184
 4 CD-1         3E                   2.9  rep2      female     0.00187
 5 CD-1         2B                   2.9  rep2      female     0.00178
 6 CD-1         1A                   2.9  rep3      female     0.00188
 7 BALB C       2B                   2.9  rep1      female     0.00187
 8 CD-1         3E                   2.92 rep3      female     0.00193
 9 CD-1         2B                   2.98 rep1      &lt;NA&gt;       0.00179
10 CD-1         2B                   3    rep1      female     0.00185
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>If we compare this to when we just printed our data above, we can see that the rows are now sorted so that the mice that lost the least weight are at the top.</p>
<p>Sometimes you might want to sort in descending order instead (largest values first). You can do this by putting the <code>desc()</code> function around your column name, inside <code>arrange()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># put desc() around the column name to sort in descending order</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 Black 6      3E                   6.3  rep1      male      0.00221
 2 Black 6      3E                   6.05 rep1      male      0.0023 
 3 Black 6      3E                   6    rep2      male      0.0022 
 4 Black 6      3E                   6    rep3      male      0.00222
 5 Black 6      3E                   5.95 rep2      male      0.00223
 6 Black 6      3E                   5.95 rep3      male      0.00229
 7 Black 6      3E                   5.85 rep1      male      0.00213
 8 Black 6      3E                   5.85 rep1      male      0.00217
 9 Black 6      3E                   5.85 rep3      male      0.0023 
10 Black 6      3E                   5.8  rep2      male      0.00229
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Now we can see the mice that lost the most weight are at the top.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Comments and pipes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Comments and pipes
</div>
</div>
<div class="callout-body-container callout-body">
<p>Notice how in the previous example we have written a comment in the middle of the pipe chain. This is a good practice to help you remember what each step is doing, especially when you have a long chain of functions, and won’t cause any errors as long as you make sure that the comment is on its own line.</p>
<p>You can also write comments at the end of the line, just make sure it’s after the pipe operator <code>%&gt;%</code>.</p>
<p>For example, these comments are allowed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span> <span class="co"># a comment here is fine</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># a comment here is fine</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
 1 Black 6      3E                   6.3  rep1      male      0.00221
 2 Black 6      3E                   6.05 rep1      male      0.0023 
 3 Black 6      3E                   6    rep2      male      0.0022 
 4 Black 6      3E                   6    rep3      male      0.00222
 5 Black 6      3E                   5.95 rep2      male      0.00223
 6 Black 6      3E                   5.95 rep3      male      0.00229
 7 Black 6      3E                   5.85 rep1      male      0.00213
 8 Black 6      3E                   5.85 rep1      male      0.00217
 9 Black 6      3E                   5.85 rep3      male      0.0023 
10 Black 6      3E                   5.8  rep2      male      0.00229
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>But this will cause an error, because the <code>#</code> is before the pipe, so R treats it as part of the comment (notice how the <code>%&gt;%</code> has changed colour?) and doesn’t know how the two lines relate to each other. It tries to run them separately, which for the first line is ok (it will just print <code>m_dose</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="co"># this comment will cause an error %&gt;%</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'weight_lost_g' not found</code></pre>
</div>
</div>
<p>But for the second line, there is an error that R doesn’t know what the <code>weight_lost_g</code> object is. That’s because it’s a column in the <code>m_dose</code> data frame, so R only knows what it is in the context of the pipe chain containing that data frame.</p>
</div>
</div>
<p>You can also sort by multiple columns by passing multiple column names to <code>arrange()</code>. For example, to sort by the strain first and then by the amount of weight lost:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 BALB C       2B                   2.7  rep2      female     0.00192
 2 BALB C       2B                   2.9  rep1      female     0.00187
 3 BALB C       2B                   3.2  rep2      female     0.00187
 4 BALB C       2B                   3.25 rep1      female     0.00178
 5 BALB C       2B                   3.25 rep3      male       0.00187
 6 BALB C       2B                   3.25 rep3      female     0.00191
 7 BALB C       2B                   3.3  rep1      male       0.00197
 8 BALB C       2B                   3.3  rep1      female     0.00195
 9 BALB C       2B                   3.32 rep3      female     0.00199
10 BALB C       2B                   3.35 rep2      female     0.00187
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This will sort the data frame by strain (according to alphabetical order, as it is a character column), and within each strain, they are then sorted by the amount of weight lost.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Piping into View()">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Piping into View()
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the above example, we sorted the data by strain and then by weight lost, but because there are so many mice in each strain, the preview shown in our console doesn’t allow us to see the full effect of the sorting.</p>
<p>One handy trick you can use with pipes is to add <code>View()</code> at the end of your chain to open the data in a separate window. Try running this code, and you’ll be able to scroll through the full dataset to check that the other mouse strains have also been sorted correctly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">View</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is a great way to check that your code has actually done what you intended!</p>
</div>
</div>
<section id="sec-sliceMinMax" class="level4" data-number="2.2.1.1">
<h4 data-number="2.2.1.1" class="anchored" data-anchor-id="sec-sliceMinMax"><span class="header-section-number">2.2.1.1</span> Extracting rows with the smallest or largest values</h4>
<p>Slice functions are used to select rows based on their position in the data frame. The <code>slice_min()</code> and <code>slice_max()</code> functions are particularly useful, because they allow you to select the rows with the smallest or largest values in a particular column.</p>
<p>This is equivalent to using <code>arrange()</code> followed by <code>head()</code>, but is more concise:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the 10 mice with the lowest drug dose</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_min() requires the column to sort by, and n = the number of rows to keep</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(drug_dose_g, <span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 13 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         3E                   3.15 rep1      female     0.00172
 2 CD-1         3E                   3.4  rep1      female     0.00174
 3 CD-1         1A                   3.45 rep3      female     0.00176
 4 CD-1         2B                   3.25 rep1      female     0.00178
 5 CD-1         2B                   3.9  rep1      male       0.00178
 6 CD-1         2B                   2.9  rep2      female     0.00178
 7 BALB C       2B                   3.25 rep1      female     0.00178
 8 CD-1         2B                   2.98 rep1      &lt;NA&gt;       0.00179
 9 CD-1         1A                   3.7  rep1      &lt;NA&gt;       0.0018 
10 CD-1         3E                   3.6  rep1      male       0.0018 
11 CD-1         3E                   3.8  rep1      male       0.0018 
12 CD-1         3E                   3.95 rep1      male       0.0018 
13 CD-1         2B                   3.55 rep1      female     0.0018 
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_max() has the same arguments as slice_min()</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
3 Black 6      3E                   6    rep2      male      0.0022 
4 Black 6      3E                   6    rep3      male      0.00222
5 Black 6      3E                   5.95 rep2      male      0.00223
6 Black 6      3E                   5.95 rep3      male      0.00229
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>But wait— neither of those pieces of code actually gave the number of rows we asked for! In the first example, we asked for the 10 mice with the lowest drug dose, but we got 13. And in the second example, we asked for the top 5 mice that lost the most weight, but we got 6. Why aren’t the <code>slice_</code> functions behaving as expected?</p>
<p>If we take a look at the help page (type <code>?slice_min</code> in the console), we learn that <code>slice_min()</code> and <code>slice_max()</code> have an argument called <code>with_ties</code> that is set to <code>TRUE</code> by default. If we want to make sure we only get the number of rows we asked for, we would have to set it to <code>FALSE</code>, like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no ties allowed!</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
3 Black 6      3E                   6    rep2      male      0.0022 
4 Black 6      3E                   6    rep3      male      0.00222
5 Black 6      3E                   5.95 rep2      male      0.00223
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This is an important lesson: sometimes functions will behave in a way that is unexpected, and you might need to read their help page or use other guides/google/AI to understand why.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which code would you use to sort the <code>m_dose</code> data frame from biggest to smallest initial weight?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% sort(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% arrange(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% sort(descending(initial_weight_g))</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% arrange(desc(initial_weight_g))</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Which code would you use to extract the 3 mice with the highest initial weight from the <code>m_dose</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% slice_max(initial_weight_g, n = 3)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% arrange(desc(initial_weight_g))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% slice_min(initial_weight_g, n = 3)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% arrange(initial_weight_g)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. I’ve written the below code, but one of the comments is messing it up! Which one?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># comment A</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="co"># comment B %&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># comment C</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>) <span class="co"># comment D</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Comment A</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Comment B</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Comment C</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>Comment D</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The correct code to sort the <code>m_dose</code> data frame from biggest to smallest initial weight is <code>m_dose %&gt;% arrange(desc(initial_weight_g))</code>. The <code>arrange()</code> function is used to sort the data frame (although there is a <code>sort()</code> function in R, that’s not part of dplyr and won’t work the same way), and the <code>desc()</code> function is used to sort in descending order.</p></li>
<li><p>The correct code to extract the 3 mice with the highest initial weight from the <code>m_dose</code> data frame is <code>m_dose %&gt;% slice_max(initial_weight_g, n = 3)</code>. The <code>slice_max()</code> function is used to select the rows with the largest values in the <code>initial_weight_g</code> column, and the <code>n = 3</code> argument specifies that we want to keep 3 rows. The <code>arrange()</code> function is not needed in this case, because <code>slice_max()</code> will automatically sort the data frame by the specified column.</p></li>
<li><p>The comment that is messing up the code is Comment B. The <code>#</code> symbol is before the pipe operator <code>%&gt;%</code>, so R treats it as part of the comment and this breaks our chain of pipes. The other comments are fine, because they are either at the end of the line or on their own line. Basically, if a comment is changing the colour of the pipe operator (or any other bits of your code), it’s in the wrong place!</p></li>
</ol>
</details>
</div>
</div>
</section>
</section>
<section id="sec-filter" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="sec-filter"><span class="header-section-number">2.2.2</span> Filtering data (rows)</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F03_filter.png" class="img-fluid figure-img"></p>
<figcaption>Filter allows you to filter rows using a logical test</figcaption>
</figure>
</div>
<p>In <code>dplyr</code>, the <code>filter()</code> function is used to subset rows based on their values. You provide a logical test, and <code>filter()</code> will keep the rows where the test is <code>TRUE</code>. We can write these tests using the comparison operators we learned in the previous chapter (e.g.&nbsp;<code>==</code>, <code>&lt;</code> and <code>!=</code>, see <a href="chapter_1.html#sec-comparisons" class="quarto-xref">Section&nbsp;<span>1.3</span></a>).</p>
<p>For example, to filter the <code>m_dose</code> data frame to only include mice that lost more than 6g:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 Black 6      3E                   6.3  rep1      male      0.00221
2 Black 6      3E                   6.05 rep1      male      0.0023 
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or to only include mice from cage 3E:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that == is used for testing equality</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="co"># don't forget the quotes either!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 168 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         3E                   3.4  rep1      female     0.00174
 2 CD-1         3E                   3.6  rep1      male       0.0018 
 3 CD-1         3E                   3.8  rep1      female     0.00189
 4 CD-1         3E                   3.95 rep1      male       0.00185
 5 CD-1         3E                   3.8  rep1      male       0.0018 
 6 CD-1         3E                   3.8  rep1      female     0.00187
 7 CD-1         3E                   3.55 rep1      male       0.00183
 8 CD-1         3E                   3.2  rep1      female     0.00187
 9 CD-1         3E                   3.15 rep1      female     0.00172
10 CD-1         3E                   3.95 rep1      male       0.0018 
# ℹ 158 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<section id="combining-logical-tests" class="level4" data-number="2.2.2.1">
<h4 data-number="2.2.2.1" class="anchored" data-anchor-id="combining-logical-tests"><span class="header-section-number">2.2.2.1</span> Combining logical tests</h4>
<p>Sometimes we want to filter based on multiple conditions. Here we will show some more advanced operators that can be used to combine logical tests.</p>
<p>The <code>&amp;</code> operator is used to combine two logical tests with an ‘and’ condition. For example, to filter the data frame to only include mice that have a tail length greater than 19mm and are female:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(tail_length_mm <span class="sc">&gt;</span> <span class="dv">19</span> <span class="sc">&amp;</span> sex <span class="sc">==</span> <span class="st">"female"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 9
  mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
1 CD-1         1A                   3.45 rep1      female     0.00193
2 CD-1         3E                   3.8  rep1      female     0.00189
3 CD-1         2B                   3.3  rep1      female     0.00181
4 CD-1         3E                   3.9  rep3      female     0.00191
5 BALB C       2B                   3.52 rep3      female     0.00194
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The <code>|</code> operator is used to combine two logical tests with an ‘or’ condition. For example, to filter the data frame to only include mice that have an initial weight less than 35g or a tail length less than 14mm:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(initial_weight_g <span class="sc">&lt;</span> <span class="dv">35</span> <span class="sc">|</span> tail_length_mm <span class="sc">&lt;</span> <span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
 2 CD-1         1A                   4.4  rep1      male       0.00198
 3 CD-1         1A                   3.32 rep1      female     0.00184
 4 CD-1         3E                   2.9  rep2      female     0.00187
 5 CD-1         1A                   3.6  rep2      female     0.0019 
 6 CD-1         1A                   3.2  rep2      female     0.00189
 7 CD-1         2B                   3.4  rep2      female     0.00185
 8 CD-1         2B                   2.9  rep2      female     0.00178
 9 CD-1         2B                   3.05 rep3      female     0.00188
10 Black 6      3E                   4.5  rep1      female     0.00211
# ℹ 20 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The <code>%in%</code> operator can be used to filter based on a vector of multiple values (<code>c(x, y)</code>). It’s particularly useful when you have a few character values you want to filter on, as it is shorter to type than <code>|</code> (or).</p>
<p>For example, to filter the data frame to only include mice from cages 3E or 1A, we could use <code>|</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span> <span class="sc">|</span> cage_number <span class="sc">==</span> <span class="st">"1A"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 220 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 210 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or we could use <code>%in%</code> like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"3E"</span>, <span class="st">"1A"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 220 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 210 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which code would you use to filter the <code>m_dose</code> data frame to only include mice from replicate 2?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(replicate == 2)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(replicate == rep2)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% filter(replicate == "rep2")</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(replicate = "rep2")</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. What is NOT a valid way to filter the <code>m_dose</code> data frame to only include mice that lost more than 4g, and have an initial weight less than 40g?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(weight_lost_g &gt; 4) %&gt;% filter(initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(weight_lost_g &gt; 4 &amp; initial_weight_g &lt; 40)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(initial_weight_g &lt; 40) %&gt;% filter(weight_lost_g &gt; 4)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. Which option correctly describes what the following code is doing?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(mouse_strain <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"BALB C"</span>, <span class="st">"Black 6"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">3</span> <span class="sc">&amp;</span> weight_lost_g <span class="sc">&lt;</span> <span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to remove mice from the “BALB C” and “Black 6” strains, who only lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to remove mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span>Filters the data frame to only include mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span>Filters the data frame to only include mice from the “BALB C” and “Black 6” strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The correct code to filter the <code>m_dose</code> data frame to only include mice from replicate 2 is <code>m_dose %&gt;% filter(replicate == "rep2")</code>. Option A is incorrect because <code>2</code> is not a value of <code>replicate</code> (when filtering you need to know what values are actually in your columns! So make sure to <code>View()</code> your data first). Option B is incorrect because the replicate column is a character column, so you need to use quotes around the value you are filtering on. Option D is incorrect because <code>=</code> is not the correct way to test for equality, you need to use <code>==</code>.</p></li>
<li><p>The invalid option is <code>m_dose %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)</code>. This is because the second filtering step is missing the name of the filter function, so R doesn’t know what to do with <code>(initial_weight_g &lt; 40)</code>. The other options are valid ways to filter the data frame based on the specified conditions; note that we can use multiple <code>filter()</code> functions in a row to apply multiple conditions, or the <code>&amp;</code> operator to combine them into a single <code>filter()</code> function. It’s just a matter of personal preference.</p></li>
<li><p>The correct description of the code is that it filters the data frame to only include mice from the “BALB C” and “Black 6” strains, then filters those further to only those that lost between 3 and 5g of weight, and finally sorts the data frame by drug dose in descending order.</p></li>
</ol>
</details>
</div>
</div>
</section>
</section>
<section id="sec-missing" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="sec-missing"><span class="header-section-number">2.2.3</span> Dealing with missing values</h3>
<p>Missing values are a common problem in real-world datasets. In R, missing values are represented by <code>NA</code>. In fact, if you look at the <code>m_dose</code> data frame we’ve been using, you’ll see that some of the cells contain <code>NA</code>: try spotting them with the <code>View()</code> function.</p>
<p>You can also find missing values in a data frame using the <code>is.na()</code> function in combination with <code>filter()</code>. For example, to find all the rows in the <code>m_dose</code> data frame that have a missing value for the <code>drug_dose_g</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 9
  mouse_strain cage_number weight_lost_g replicate sex   drug_dose_g
  &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;
1 CD-1         1A                     NA rep1      &lt;NA&gt;           NA
2 Black 6      3E                     NA rep3      &lt;NA&gt;           NA
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The problem with missing values is that they can cause problems when you try to perform calculations on your data. For example, if you try to calculate the mean of a column that contains even a single missing value, the result will also be <code>NA</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from chapter 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>m_dose<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
</div>
<p><code>NA</code> values in R are therefore referred to as ‘contagious’: if you put an <code>NA</code> in you usually get an <code>NA</code> out. If you think about it, that makes sense— when we don’t know the value of a particular mouse’s drug dose, how can we calculate the average? That missing value could be anything.</p>
<p>For this reason, it’s important to deal with missing values before performing calculations. Many functions in R will have an argument called <code>na.rm</code> that you can set to <code>TRUE</code> to remove missing values before performing the calculation. For example, to calculate the mean of the <code>drug_dose_g</code> column with the missing values excluded:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from chapter 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>m_dose<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.002009152</code></pre>
</div>
</div>
<p>This time, the result is a number, because the missing values have been removed before the calculation.</p>
<p>But not all functions have an <code>na.rm</code> argument. In these cases, you can remove rows with missing values. This can be done for a single column, using the <code>filter()</code> function together with <code>is.na()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in the drug_dose_g column</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember the ! means 'not', it negates the result of is.na()</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(drug_dose_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 342 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
 9 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
10 CD-1         1A                   3.3  rep1      &lt;NA&gt;       0.00186
# ℹ 332 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Or, you can remove rows with missing values in any column using the <code>na.omit()</code> or <code>drop_na()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in any column</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 333 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.2  rep1      female     0.00182
 9 CD-1         1A                   3.8  rep1      male       0.00191
10 CD-1         1A                   4.4  rep1      male       0.00198
# ℹ 323 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 333 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.2  rep1      female     0.00182
 9 CD-1         1A                   3.8  rep1      male       0.00191
10 CD-1         1A                   4.4  rep1      male       0.00198
# ℹ 323 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Sometimes, instead of removing rows with missing values, you might want to replace them with a specific value. This can be done using the <code>replace_na()</code> function from the <code>tidyr</code> package. <code>replace_na()</code> takes a <code>list()</code> which contains each of the column names you want to edit, and the value that should be used.</p>
<p>For example, to replace missing values in the <code>weight_lost_g</code> columns with 0, replace missing values in the <code>sex</code> column with ‘unknown’ and leave the rest of the data frame unchanged:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># replace missing values in the drug_dose_g column with 0</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we need to provide the column_names = values_to_replace</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this needs to be contained within a list()</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">weight_lost_g =</span> <span class="dv">0</span>, <span class="at">sex =</span> <span class="st">"unknown"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex     drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;         &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male        0.00181
 2 CD-1         1A                   3.8  rep1      female      0.00186
 3 CD-1         1A                   3.25 rep1      female      0.00195
 4 CD-1         1A                   0    rep1      unknown    NA      
 5 CD-1         1A                   3.45 rep1      female      0.00193
 6 CD-1         1A                   3.65 rep1      male        0.0019 
 7 CD-1         1A                   3.62 rep1      female      0.00181
 8 CD-1         1A                   4.68 rep1      male        0.00195
 9 CD-1         1A                   3.48 rep1      unknown     0.00193
10 CD-1         1A                   4.25 rep1      unknown     0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>When deciding how to handle missing values, you might have prior knowledge that <code>NA</code> should be replaced with a specific value, or you might decide that removing rows with <code>NA</code> is the best approach for your analysis.</p>
<p>For example, maybe we knew that the mice were given a <code>weight_lost_g</code> of <code>NA</code> if they didn’t lose any weight, it would then make sense to replace those with 0 (as we did in the code above). However, if the <code>drug_dose_g</code> column was missing simply because the data was lost, we might choose to remove those rows entirely.</p>
<p>It’s important to think carefully about how missing values should be handled in your analysis.</p>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. What would be the result of running this R code: <code>mean(c(1, 2, 4, NA))</code></p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span>2.333333</p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span>0</p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>NA</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span>An error</p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. Which line of code would you use to filter the <code>m_dose</code> data frame to remove mice that have a missing value in the <code>tail_length_mm</code> column?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(tail_length_mm != NA)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(is.na(tail_length_mm))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% na.omit()</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% filter(!is.na(tail_length_mm))</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. How would you replace missing values in the <code>initial_weight_g</code> column with the value 35?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% replace_na(list(initial_weight_g = 35))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% replace_na(initial_weight_g = 35)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% replace_na(list(initial_weight_g == 35))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% replace_na(35)</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The result of running the code <code>mean(c(1, 2, 4, NA))</code> is <code>NA</code>. This is because the <code>NA</code> value is ‘contagious’, so when you try to calculate the mean of a vector that contains an <code>NA</code>, the result will also be <code>NA</code>. If we wanted to calculate the mean of the vector without the <code>NA</code>, we would need to use the <code>na.rm = TRUE</code> argument.</li>
<li>The correct line of code to filter the <code>m_dose</code> data frame to remove mice that have a missing value in the <code>tail_length_mm</code> column is <code>m_dose %&gt;% filter(!is.na(tail_length_mm))</code>. The <code>!</code> symbol is used to negate the result of <code>is.na()</code>, so we are filtering to keep the rows where <code>tail_length_mm</code> is not <code>NA</code>. We can’t use the first option with the <code>!= NA</code> because <code>NA</code> is a special value in R that represents missing data, and it can’t be compared to anything, and the third option is incorrect because <code>na.omit()</code> removes entire rows with missing values, rather than just filtering based on a single column.</li>
<li>The correct line of code to replace missing values in the <code>initial_weight_g</code> column with the value 35 is <code>m_dose %&gt;% replace_na(list(initial_weight_g = 35))</code>. The <code>replace_na()</code> function takes a <code>list()</code> that contains the column names you want to replace and the values you want to replace them with. We only need to use a single equal sign here as we’re not testing for equality, we’re assigning a value.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
<section id="sec-select" class="level3" data-number="2.2.4">
<h3 data-number="2.2.4" class="anchored" data-anchor-id="sec-select"><span class="header-section-number">2.2.4</span> Selecting columns</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F04_select.png" class="img-fluid figure-img"></p>
<figcaption>Select allows you to select only certain columns</figcaption>
</figure>
</div>
<p>While <code>filter()</code> is used to subset rows, <code>select()</code> is used to subset columns. You can use <code>select()</code> to keep only the columns you’re interested in, or to drop columns you don’t need.</p>
<p>The <code>select()</code> function takes the names of the columns that you want to keep/remove (no vector notation <code>c()</code> or quotation marks <code>""</code> necessary). For example, to select only the <code>mouse_strain</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>m_dose</code> data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, initial_weight_g, weight_lost_g)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 3
   mouse_strain initial_weight_g weight_lost_g
   &lt;chr&gt;                   &lt;dbl&gt;         &lt;dbl&gt;
 1 CD-1                     39.1          3.75
 2 CD-1                     39.5          3.8 
 3 CD-1                     40.3          3.25
 4 CD-1                     NA           NA   
 5 CD-1                     36.7          3.45
 6 CD-1                     39.3          3.65
 7 CD-1                     38.9          3.62
 8 CD-1                     39.2          4.68
 9 CD-1                     34.1          3.48
10 CD-1                     42            4.25
# ℹ 334 more rows</code></pre>
</div>
</div>
<p>We can see that all the other columns have been removed from the data frame.</p>
<p>If you want to keep all columns except for a few, you can use <code>-</code> to drop columns. For example, to keep all columns except for <code>cage_number</code> and <code>sex</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>cage_number, <span class="sc">-</span>sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 7
   mouse_strain weight_lost_g replicate drug_dose_g tail_length_mm
   &lt;chr&gt;                &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;          &lt;dbl&gt;
 1 CD-1                  3.75 rep1          0.00181           18.7
 2 CD-1                  3.8  rep1          0.00186           17.4
 3 CD-1                  3.25 rep1          0.00195           18  
 4 CD-1                 NA    rep1         NA                 NA  
 5 CD-1                  3.45 rep1          0.00193           19.3
 6 CD-1                  3.65 rep1          0.0019            20.6
 7 CD-1                  3.62 rep1          0.00181           17.8
 8 CD-1                  4.68 rep1          0.00195           19.6
 9 CD-1                  3.48 rep1          0.00193           18.1
10 CD-1                  4.25 rep1          0.0019            20.2
# ℹ 334 more rows
# ℹ 2 more variables: initial_weight_g &lt;dbl&gt;, id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>There are also some helper functions that can be used to select columns based on their names :</p>
<table class="caption-top table">
<caption>There are several helper functions that can be used with the select function</caption>
<colgroup>
<col style="width: 14%">
<col style="width: 42%">
<col style="width: 41%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Description</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>starts_with()</code></td>
<td>select column(s) that start with a certain string</td>
<td><p>select all columns starting with the letter i</p>
<p><code>select(starts_with("i"))</code></p></td>
</tr>
<tr class="even">
<td><code>ends_with()</code></td>
<td>select column(s) that end with a certain string</td>
<td><p>select all columns ending with _g</p>
<p><code>select(ends_with("_g"))</code></p></td>
</tr>
<tr class="odd">
<td><code>contains()</code></td>
<td>select column(s) that contain a certain string</td>
<td><p>select all columns containing the word ‘weight’</p>
<p><code>select(contains("weight"))</code></p></td>
</tr>
</tbody>
</table>
<p>You need to use quotation marks around the arguments in these helper functions, as they aren’t full column names, just strings of characters.</p>
<p>Try using these helper functions to select columns from the <code>m_dose</code> data frame!</p>
<div class="callout callout-style-default callout-note callout-titled" title="Reordering columns">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Reordering columns
</div>
</div>
<div class="callout-body-container callout-body">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F05_reorder.png" class="img-fluid figure-img"></p>
<figcaption>Relocate allows you to move columns around</figcaption>
</figure>
</div>
<p>We can reorder columns using the <code>relocate()</code> function, which works similarly to <code>select()</code> (except it just moves columns around rather than dropping/keeping them). For example, to move the <code>sex</code> column to before the <code>cage_number</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first the name of the column to move, then where it should go</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(sex, <span class="at">.before =</span> cage_number)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain sex    cage_number weight_lost_g replicate drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;
 1 CD-1         male   1A                   3.75 rep1          0.00181
 2 CD-1         female 1A                   3.8  rep1          0.00186
 3 CD-1         female 1A                   3.25 rep1          0.00195
 4 CD-1         &lt;NA&gt;   1A                  NA    rep1         NA      
 5 CD-1         female 1A                   3.45 rep1          0.00193
 6 CD-1         male   1A                   3.65 rep1          0.0019 
 7 CD-1         female 1A                   3.62 rep1          0.00181
 8 CD-1         male   1A                   4.68 rep1          0.00195
 9 CD-1         &lt;NA&gt;   1A                   3.48 rep1          0.00193
10 CD-1         &lt;NA&gt;   1A                   4.25 rep1          0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Without a specific position <code>( .before / .after)</code>, this function will place the chosen column(s) as the first / left-most columns.</p>
<p>Two further useful helper functions for <code>relocate()</code> are the <code>everything()</code> and <code>last_col()</code> functions, which can be used to move columns to the start/end of the data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move id_num to the front</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(id_num, <span class="at">.before =</span> <span class="fu">everything</span>()) <span class="co"># don't forget the brackets</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   id_num mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
    &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1      1 CD-1         1A                   3.75 rep1      male       0.00181
 2      2 CD-1         1A                   3.8  rep1      female     0.00186
 3      3 CD-1         1A                   3.25 rep1      female     0.00195
 4      4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5      5 CD-1         1A                   3.45 rep1      female     0.00193
 6      6 CD-1         1A                   3.65 rep1      male       0.0019 
 7      7 CD-1         1A                   3.62 rep1      female     0.00181
 8      8 CD-1         1A                   4.68 rep1      male       0.00195
 9      9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10     10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 2 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move mouse_strain to the end</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(mouse_strain, <span class="at">.after =</span> <span class="fu">last_col</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   cage_number weight_lost_g replicate sex    drug_dose_g tail_length_mm
   &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;
 1 1A                   3.75 rep1      male       0.00181           18.7
 2 1A                   3.8  rep1      female     0.00186           17.4
 3 1A                   3.25 rep1      female     0.00195           18  
 4 1A                  NA    rep1      &lt;NA&gt;      NA                 NA  
 5 1A                   3.45 rep1      female     0.00193           19.3
 6 1A                   3.65 rep1      male       0.0019            20.6
 7 1A                   3.62 rep1      female     0.00181           17.8
 8 1A                   4.68 rep1      male       0.00195           19.6
 9 1A                   3.48 rep1      &lt;NA&gt;       0.00193           18.1
10 1A                   4.25 rep1      &lt;NA&gt;       0.0019            20.2
# ℹ 334 more rows
# ℹ 3 more variables: initial_weight_g &lt;dbl&gt;, id_num &lt;dbl&gt;, mouse_strain &lt;chr&gt;</code></pre>
</div>
</div>
<p>Re-ordering columns isn’t necessary, but it makes it easier to see the data you’re most interested in within the console (since often not all of the columns will fit on the screen at once). For example, if we are doing a lot of computation on the <code>initial_weight_g</code> column, we’d probably like to have that near the start so we can easily check it.</p>
</div>
</div>
<p>Note that the output of the <code>select()</code> function is a new data frame, even if you only select a single column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select the mouse_strain column</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># recall from chapter 1 that class() tells us the type of an object</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "tbl_df"     "tbl"        "data.frame"</code></pre>
</div>
</div>
<p>Sometimes, we instead want to get the values of a column as a vector.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F06_pull.png" class="img-fluid figure-img"></p>
<figcaption>Pull allows you to pull a column out of a data frame as a vector</figcaption>
</figure>
</div>
<p>We can do this by using the <code>pull()</code> function, which extracts a single column from a data frame as a vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the mouse_strain column as a vector</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
</div>
<p>We can see that the class of the output is now a vector, rather than a data frame. This is important because some functions only accept vectors, not data frames, like <code>mean()</code> for example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this will give an error</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span> <span class="fu">select</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in mean.default(., na.rm = TRUE): argument is not numeric or logical:
returning NA</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this will work</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span> <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 43.92193</code></pre>
</div>
</div>
<p>Note how both times we used <code>na.rm = TRUE</code> to remove missing values before calculating the mean.</p>
<p>You might remember that we used the <code>$</code> operator in the previous chapter to extract a single column from a data frame, so why use <code>pull()</code> instead? The main reason is that <code>pull()</code> works within a chain of pipes, whereas <code>$</code> doesn’t.</p>
<p>For example, let’s say we want to know the average initial weight of mice that lost at least 4g. We can do this by chaining <code>filter()</code> and <code>pull()</code> together:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter to mice that lost at least 4g</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;=</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the initial_weight_g column as a vector</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate mean, removing NA values</span></span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 46.48023</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Practice exercises">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try these practice questions to test your understanding</p>
<div class="question">
<p>1. Which line of code would NOT be a valid way to select only the <code>drug_dose_g</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>m_dose</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% select(drug_dose_g, initial_weight_g, weight_lost_g)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% select(contains("g"))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% select(ends_with("_g"))</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% select(-cage_number, -tail_length_mm, -id_num, -mouse_strain, -sex, -replicate)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>2. How would I extract the <code>initial_weight_g</code> column from the <code>m_dose</code> data frame as a vector?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% filter(initial_weight_g)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% $initial_weight_g</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no3"> <span id="no3" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% select(initial_weight_g)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% pull(initial_weight_g)</code></p><p></p>
</div>
</div>

</div>
</div>
<div class="question">
<p>3. How would you move the <code>sex</code> column to the end of the <code>m_dose</code> data frame?</p>
<div class="choices">
<div x-data="{ answer: '' , clear() { this.answer = '' }}">

<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no1"> <span id="no1" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% relocate(sex)</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no2"> <span id="no2" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% relocate(sex, .after = last_col)</code></p><p></p>
</div>
<div class="choice correct-choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="correct"> <span id="correct" x-show="answer == $el.id" class="badge">✔</span><code>m_dose %&gt;% relocate(sex, .after = last_col())</code></p><p></p>
</div>
<div class="choice">
<p></p><p><input class="form-check-input" type="radio" x-model="answer" value="no4"> <span id="no4" x-show="answer == $el.id" class="badge">✗</span><code>m_dose %&gt;% reorder(sex, .after = last_col())</code></p><p></p>
</div>
</div>

</div>
</div>
<details>
<summary>
Solutions
</summary>
<p>
</p><ol type="1">
<li>The line of code that would NOT be a valid way to select the <code>drug_dose_g</code>, <code>initial_weight_g</code>, and <code>weight_lost_g</code> columns from the <code>m_dose</code> data frame is <code>m_dose %&gt;% select(contains("g"))</code>. This line of code would select all columns that contain the letter ‘g’, which would include columns like <code>cage_number</code> and <code>tail_length_mm</code>. We need to specify either <code>ends_with("g")</code> or <code>contains("_g")</code> to only get those with <code>_g</code> at the end. The other options are valid ways to select the specified columns, although some are more efficient than others!</li>
<li>The correct way to extract the <code>initial_weight_g</code> column from the <code>m_dose</code> data frame as a vector is <code>m_dose %&gt;% pull(initial_weight_g)</code>. The <code>pull()</code> function is used to extract a single column from a data frame as a vector. The other options are incorrect because <code>filter()</code> is used to subset rows, <code>$</code> is not used in a pipe chain, and <code>select()</code> is outputs a data frame, not extract them as vectors.</li>
<li>The correct way to move the <code>sex</code> column to the end of the <code>m_dose</code> data frame is using the <code>relocate()</code> function like this: <code>m_dose %&gt;% relocate(sex, .after = last_col())</code>. The <code>last_col()</code> function is used to refer to the last column in the data frame. The other options are incorrect because <code>reorder()</code> is not a valid function, and you need to remember to include the brackets <code>()</code> when using <code>last_col()</code>.</li>
</ol>
<p></p>
</details>
</div>
</div>
</section>
</section>
<section id="summary" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="summary"><span class="header-section-number">2.3</span> Summary</h2>
<p>Here’s what we’ve covered in this chapter:</p>
<ul>
<li><p>The pipe operator <code>%&gt;%</code> and how we can use it to chain together multiple function calls, making our code more readable and easier to understand.</p></li>
<li><p>The basic dplyr verbs <code>arrange()</code>, <code>filter()</code> and <code>select()</code></p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled" title="Why does data need to be tidy anyway?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why does data need to be tidy anyway?
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this chapter, we’ve been focusing on making our data ‘tidy’: that is, structured in a consistent way that makes it easy to work with. A nice visual illustration of tidy data and its importance can be <a href="https://allisonhorst.com/other-r-fun">found here</a>.</p>
</div>
</div>
<section id="practice-questions" class="level3" data-number="2.3.1">
<h3 data-number="2.3.1" class="anchored" data-anchor-id="practice-questions"><span class="header-section-number">2.3.1</span> Practice questions</h3>
<ol type="1">
<li><p>What is the purpose of the pipe operator <code>%&gt;%</code>? Keeping this in mind, re-write the following code to use the pipe.</p>
<ol type="a">
<li><code>round(mean(c(1, 2, 3, 4, 5)))</code></li>
<li><code>print(as.character(1 + 10))</code></li>
</ol></li>
<li><p>What would be the result of evaluating the following expressions? You don’t need to know these off the top of your head, use R to help! (Hint: some expressions might give an error. Try to think about why)</p>
<ol type="a">
<li><code>m_dose %&gt;% filter(weight_lost_g &gt; 10)</code></li>
<li><code>m_dose %&gt;% select(tail_length_mm, weight_lost_g)</code></li>
<li><code>m_dose %&gt;% arrange(tail_length_mm)</code></li>
<li><code>m_dose %&gt;% filter(initial_Weight_g &gt; 10) %&gt;% arrange(mouse_strain)</code></li>
<li><code>m_dose %&gt;% relocate(mouse_strain, .after = cage_number)</code></li>
<li><code>m_dose %&gt;% pull(weight_lost_g)</code></li>
<li><code>m_dose %&gt;% filter(!is.na(weight_lost_g))</code></li>
<li><code>m_dose %&gt;% replace_na(list(weight_lost_g = 0))</code></li>
</ol></li>
<li><p>What is a missing value in R? What are two ways to deal with missing values in a data frame?</p></li>
</ol>
<details>
<summary>
Solutions
</summary>
<ol type="1">
<li><p>The pipe operator <code>%&gt;%</code> is used to chain together multiple function calls, passing the result of one function to the next. Here’s how you could re-write the code to use the pipe:</p>
<ol type="a">
<li><code>c(1, 2, 3, 4, 5) %&gt;% mean() %&gt;% round()</code></li>
<li><code>as.character(1 + 10) %&gt;% print()</code></li>
</ol></li>
<li><p>The result of evaluating the expressions would be:</p>
<ol type="a">
<li>A data frame containing only the rows where <code>weight_lost_g</code> is greater than 10.</li>
<li>A data frame containing only the <code>tail_length_mm</code> and <code>weight_lost_g</code> columns.</li>
<li>A data frame sorted by <code>tail_length_mm</code>, in ascending order.</li>
<li>An error because <code>initial_Weight_g</code> is not a column in the data frame.</li>
<li>A data frame with the <code>mouse_strain</code> column moved to be after the <code>cage_number</code> column.</li>
<li>A <strong>vector</strong> containing the values of the <code>weight_lost_g</code> column.</li>
<li>A data frame containing only the rows where <code>weight_lost_g</code> is not <code>NA</code>.</li>
<li>A data frame with missing values in the <code>weight_lost_g</code> column replaced with 0.</li>
</ol></li>
<li><p>A missing value in R is represented by <code>NA</code>. Two ways to deal with missing values in a data frame are to remove them using <code>filter(!is.na(column_name))</code> or to replace them with a specific value using <code>replace_na(list(column_name = value))</code>.</p></li>
</ol>
</details>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./chapter_1.html" class="pagination-link" aria-label="Introduction to R">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./chapter_3.html" class="pagination-link" aria-label="Working with data - Part 2">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Working with data - Part 2</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb82" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this chapter:"</span></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Working with data - Part 1 {#sec-chapter02}</span></span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>In this chapter we will learn how to manipulate and summarise data using the <span class="in">`dplyr`</span> package (with a little help from the <span class="in">`tidyr`</span> package too).</span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Learning Objectives"}</span>
<span id="cb82-16"><a href="#cb82-16" aria-hidden="true" tabindex="-1"></a>At the end of this chapter, learners should be able to:</span>
<span id="cb82-17"><a href="#cb82-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-18"><a href="#cb82-18" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Use the pipe (<span class="in">`%&gt;%`</span>) to chain multiple functions together</span>
<span id="cb82-19"><a href="#cb82-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-20"><a href="#cb82-20" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Design chains of dplyr functions to manipulate data frames</span>
<span id="cb82-21"><a href="#cb82-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-22"><a href="#cb82-22" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Understand how to identify and handle missing values in a data frame</span>
<span id="cb82-23"><a href="#cb82-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-24"><a href="#cb82-24" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Apply grouping for more complex analysis of data</span>
<span id="cb82-25"><a href="#cb82-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-26"><a href="#cb82-26" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Recall how to save data frames to a file</span>
<span id="cb82-27"><a href="#cb82-27" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-28"><a href="#cb82-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-29"><a href="#cb82-29" aria-hidden="true" tabindex="-1"></a>Both <span class="in">`dplyr`</span> and <span class="in">`tidyr`</span> are contained within the <span class="in">`tidyverse`</span> (along with <span class="in">`readr`</span>) so we can load all of these packages at once using <span class="in">`library(tidyverse)`</span>:</span>
<span id="cb82-30"><a href="#cb82-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-33"><a href="#cb82-33" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-34"><a href="#cb82-34" aria-hidden="true" tabindex="-1"></a><span class="co"># don't forget to load tidyverse!</span></span>
<span id="cb82-35"><a href="#cb82-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb82-36"><a href="#cb82-36" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-37"><a href="#cb82-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-38"><a href="#cb82-38" aria-hidden="true" tabindex="-1"></a><span class="fu">## Chaining functions together with pipes {#sec-pipes}</span></span>
<span id="cb82-39"><a href="#cb82-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-40"><a href="#cb82-40" aria-hidden="true" tabindex="-1"></a>Pipes are a powerful feature of the <span class="in">`tidyverse`</span> that allow you to chain multiple functions together.</span>
<span id="cb82-41"><a href="#cb82-41" aria-hidden="true" tabindex="-1"></a>Pipes are useful because they allow you to break down complex operations into smaller steps that are easier to read and understand.</span>
<span id="cb82-42"><a href="#cb82-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-43"><a href="#cb82-43" aria-hidden="true" tabindex="-1"></a>For example, take the following code:</span>
<span id="cb82-44"><a href="#cb82-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-47"><a href="#cb82-47" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-48"><a href="#cb82-48" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb82-49"><a href="#cb82-49" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(<span class="fu">round</span>(<span class="fu">mean</span>(my_vector)))</span>
<span id="cb82-50"><a href="#cb82-50" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-51"><a href="#cb82-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-52"><a href="#cb82-52" aria-hidden="true" tabindex="-1"></a>What do you think this code does?</span>
<span id="cb82-53"><a href="#cb82-53" aria-hidden="true" tabindex="-1"></a>It calculates the mean of <span class="in">`my_vector`</span>, rounds the result to the nearest whole number, and then converts the result to a character.</span>
<span id="cb82-54"><a href="#cb82-54" aria-hidden="true" tabindex="-1"></a>But the code is a bit hard to read because you have to start from the inside of the brackets and work your way out.</span>
<span id="cb82-55"><a href="#cb82-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-56"><a href="#cb82-56" aria-hidden="true" tabindex="-1"></a>Instead, we can use the pipe operator (<span class="in">`%&gt;%`</span>) to chain these functions together in a more readable way:</span>
<span id="cb82-57"><a href="#cb82-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-60"><a href="#cb82-60" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-61"><a href="#cb82-61" aria-hidden="true" tabindex="-1"></a>my_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb82-62"><a href="#cb82-62" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>() <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb82-63"><a href="#cb82-63" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-64"><a href="#cb82-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-65"><a href="#cb82-65" aria-hidden="true" tabindex="-1"></a>See how the code reads naturally from left to right?</span>
<span id="cb82-66"><a href="#cb82-66" aria-hidden="true" tabindex="-1"></a>You can think of the pipe as being like the phrase "and then".</span>
<span id="cb82-67"><a href="#cb82-67" aria-hidden="true" tabindex="-1"></a>Here, we're telling R: "Take <span class="in">`my_vector`</span>, and then calculate the mean, and then round the result, and then convert it to a character."</span>
<span id="cb82-68"><a href="#cb82-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-69"><a href="#cb82-69" aria-hidden="true" tabindex="-1"></a>You'll notice that we didn't need to specify the input to each function.</span>
<span id="cb82-70"><a href="#cb82-70" aria-hidden="true" tabindex="-1"></a>That's because the pipe automatically passes the output of the previous function as the first input to the next function.</span>
<span id="cb82-71"><a href="#cb82-71" aria-hidden="true" tabindex="-1"></a>We can still specify additional arguments to each function if we need to.</span>
<span id="cb82-72"><a href="#cb82-72" aria-hidden="true" tabindex="-1"></a>For example, if we wanted to round the mean to 2 decimal places, we could do this:</span>
<span id="cb82-73"><a href="#cb82-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-76"><a href="#cb82-76" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-77"><a href="#cb82-77" aria-hidden="true" tabindex="-1"></a>my_vector <span class="sc">%&gt;%</span> <span class="fu">mean</span>() <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb82-78"><a href="#cb82-78" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-79"><a href="#cb82-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-80"><a href="#cb82-80" aria-hidden="true" tabindex="-1"></a>R is clever enough to know that the first argument to <span class="in">`round()`</span> is still the output of <span class="in">`mean()`</span>, even though we've now specified the <span class="in">`digits`</span> argument.</span>
<span id="cb82-81"><a href="#cb82-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-82"><a href="#cb82-82" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Plenty of pipes"}</span>
<span id="cb82-83"><a href="#cb82-83" aria-hidden="true" tabindex="-1"></a>There is another style of pipe in R, called the 'base R pipe' <span class="in">`|&gt;`</span>, which is available in R version 4.1.0 and later.</span>
<span id="cb82-84"><a href="#cb82-84" aria-hidden="true" tabindex="-1"></a>The base R pipe works in a similar way to the <span class="in">`magrittr`</span> pipe (<span class="in">`%&gt;%`</span>) that we use in this course, but it is not as flexible.</span>
<span id="cb82-85"><a href="#cb82-85" aria-hidden="true" tabindex="-1"></a>We recommend using the <span class="in">`magrittr`</span> pipe for now.</span>
<span id="cb82-86"><a href="#cb82-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-87"><a href="#cb82-87" aria-hidden="true" tabindex="-1"></a>Fun fact: the <span class="in">`magrittr`</span> package is named after the <span class="co">[</span><span class="ot">artist René Magritte, who made a famous painting of a pipe</span><span class="co">](https://en.wikipedia.org/wiki/The_Treachery_of_Images)</span>.</span>
<span id="cb82-88"><a href="#cb82-88" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-89"><a href="#cb82-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-90"><a href="#cb82-90" aria-hidden="true" tabindex="-1"></a>To type the pipe operator more easily, you can use the keyboard shortcut {{&lt; kbd Cmd-shift-M &gt;}} (although once you get used to it, you might find it easier to type <span class="in">`%&gt;%`</span> manually).</span>
<span id="cb82-91"><a href="#cb82-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-92"><a href="#cb82-92" aria-hidden="true" tabindex="-1"></a>::::::::::::::: {.callout-important title="Practice exercises"}</span>
<span id="cb82-93"><a href="#cb82-93" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb82-94"><a href="#cb82-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-95"><a href="#cb82-95" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-96"><a href="#cb82-96" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span></span>
<span id="cb82-97"><a href="#cb82-97" aria-hidden="true" tabindex="-1"></a>What is NOT a valid way to re-write the following code using the pipe operator: <span class="in">`round(sqrt(sum(1:10)), 1)`</span>.</span>
<span id="cb82-98"><a href="#cb82-98" aria-hidden="true" tabindex="-1"></a>If you're not sure, try running the different options in the console to see which one gives the same answer.</span>
<span id="cb82-99"><a href="#cb82-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-100"><a href="#cb82-100" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-101"><a href="#cb82-101" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-102"><a href="#cb82-102" aria-hidden="true" tabindex="-1"></a><span class="in">`1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(1)`</span></span>
<span id="cb82-103"><a href="#cb82-103" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-104"><a href="#cb82-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-105"><a href="#cb82-105" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-106"><a href="#cb82-106" aria-hidden="true" tabindex="-1"></a><span class="in">`sum(1:10) %&gt;% sqrt(1) %&gt;% round()`</span></span>
<span id="cb82-107"><a href="#cb82-107" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-108"><a href="#cb82-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-109"><a href="#cb82-109" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-110"><a href="#cb82-110" aria-hidden="true" tabindex="-1"></a><span class="in">`1:10 %&gt;% sum() %&gt;% sqrt() %&gt;% round(digits = 1)`</span></span>
<span id="cb82-111"><a href="#cb82-111" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-112"><a href="#cb82-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-113"><a href="#cb82-113" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-114"><a href="#cb82-114" aria-hidden="true" tabindex="-1"></a><span class="in">`sum(1:10) %&gt;% sqrt() %&gt;% round(digits = 1)`</span></span>
<span id="cb82-115"><a href="#cb82-115" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-116"><a href="#cb82-116" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-117"><a href="#cb82-117" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-118"><a href="#cb82-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-119"><a href="#cb82-119" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-120"><a href="#cb82-120" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span></span>
<span id="cb82-121"><a href="#cb82-121" aria-hidden="true" tabindex="-1"></a>What is the output of the following code?</span>
<span id="cb82-122"><a href="#cb82-122" aria-hidden="true" tabindex="-1"></a><span class="in">`letters %&gt;% head() %&gt;% toupper()`</span> Try to guess it before copy-pasting into R.</span>
<span id="cb82-123"><a href="#cb82-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-124"><a href="#cb82-124" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-125"><a href="#cb82-125" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-126"><a href="#cb82-126" aria-hidden="true" tabindex="-1"></a><span class="in">`"A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z"`</span></span>
<span id="cb82-127"><a href="#cb82-127" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-128"><a href="#cb82-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-129"><a href="#cb82-129" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-130"><a href="#cb82-130" aria-hidden="true" tabindex="-1"></a><span class="in">`"a" "b" "c" "d" "e" "f"`</span></span>
<span id="cb82-131"><a href="#cb82-131" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-132"><a href="#cb82-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-133"><a href="#cb82-133" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-134"><a href="#cb82-134" aria-hidden="true" tabindex="-1"></a>An error</span>
<span id="cb82-135"><a href="#cb82-135" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-136"><a href="#cb82-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-137"><a href="#cb82-137" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-138"><a href="#cb82-138" aria-hidden="true" tabindex="-1"></a><span class="in">`"A" "B" "C" "D" "E" "F"`</span></span>
<span id="cb82-139"><a href="#cb82-139" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-140"><a href="#cb82-140" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-141"><a href="#cb82-141" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-142"><a href="#cb82-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-143"><a href="#cb82-143" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb82-144"><a href="#cb82-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-145"><a href="#cb82-145" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb82-146"><a href="#cb82-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-147"><a href="#cb82-147" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb82-148"><a href="#cb82-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-149"><a href="#cb82-149" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The invalid option is <span class="in">`sum(1:10) %&gt;% sqrt(1) %&gt;% round()`</span>. This is because the <span class="in">`sqrt()`</span> function only takes one argument, so you can't specify <span class="in">`1`</span> as an argument in addition to what is being piped in from <span class="in">`sum(1:10)`</span>. Note that some options used the pipe to send <span class="in">`1:10`</span> to <span class="in">`sum()`</span> (like <span class="in">`1:10 %&gt;% sum()`</span>), and others just used <span class="in">`sum(1:10)`</span> directly. Both are valid ways to use the pipe, it's just a matter of personal preference.</span>
<span id="cb82-150"><a href="#cb82-150" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The output of the code <span class="in">`letters %&gt;% head() %&gt;% toupper()`</span> is <span class="in">`"A" "B" "C" "D" "E" "F"`</span>. The <span class="in">`letters`</span> vector contains the lowercase alphabet, and the <span class="in">`head()`</span> function returns the first 6 elements of the vector. Finally, the <span class="in">`toupper()`</span> function then converts these elements to uppercase.</span>
<span id="cb82-151"><a href="#cb82-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-152"><a href="#cb82-152" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb82-153"><a href="#cb82-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-154"><a href="#cb82-154" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb82-155"><a href="#cb82-155" aria-hidden="true" tabindex="-1"></a>:::::::::::::::</span>
<span id="cb82-156"><a href="#cb82-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-157"><a href="#cb82-157" aria-hidden="true" tabindex="-1"></a><span class="fu">## Basic data manipulation {#sec-dataManip}</span></span>
<span id="cb82-158"><a href="#cb82-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-159"><a href="#cb82-159" aria-hidden="true" tabindex="-1"></a>To really see the power of the pipe, we will use it together with the <span class="in">`dplyr`</span> package that provides a set of functions to easily filter, sort, select, modify and summarise data frames.</span>
<span id="cb82-160"><a href="#cb82-160" aria-hidden="true" tabindex="-1"></a>These functions are designed to work well with the pipe, so you can chain them together to create complex data manipulations in a readable format.</span>
<span id="cb82-161"><a href="#cb82-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-162"><a href="#cb82-162" aria-hidden="true" tabindex="-1"></a>For example, even though we haven't covered the <span class="in">`dplyr`</span> functions yet, you can probably guess what the following code does:</span>
<span id="cb82-163"><a href="#cb82-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-166"><a href="#cb82-166" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-167"><a href="#cb82-167" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb82-168"><a href="#cb82-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-169"><a href="#cb82-169" aria-hidden="true" tabindex="-1"></a><span class="co"># use the pipe to chain together our data manipulation steps</span></span>
<span id="cb82-170"><a href="#cb82-170" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-171"><a href="#cb82-171" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb82-172"><a href="#cb82-172" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb82-173"><a href="#cb82-173" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>()</span>
<span id="cb82-174"><a href="#cb82-174" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-175"><a href="#cb82-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-176"><a href="#cb82-176" aria-hidden="true" tabindex="-1"></a>This code filters the <span class="in">`m_dose`</span> data frame to only include data from cage 3E, then pulls out the <span class="in">`weight_lost_g`</span> column, and finally calculates the mean of the values in that column.</span>
<span id="cb82-177"><a href="#cb82-177" aria-hidden="true" tabindex="-1"></a>The first argument to each function is the output of the previous function, and any additional arguments (like the column name in <span class="in">`pull()`</span>) are specified in the brackets (like <span class="in">`round(digits = 2)`</span> from the previous example).</span>
<span id="cb82-178"><a href="#cb82-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-179"><a href="#cb82-179" aria-hidden="true" tabindex="-1"></a>We also used the enter key after each pipe <span class="in">`%&gt;%`</span> to break up the code into multiple lines to make it easier to read.</span>
<span id="cb82-180"><a href="#cb82-180" aria-hidden="true" tabindex="-1"></a>This isn't required, but is a popular style in the R community, so all the code examples in this chapter will follow this format.</span>
<span id="cb82-181"><a href="#cb82-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-182"><a href="#cb82-182" aria-hidden="true" tabindex="-1"></a>We will now introduce some of the most commonly used <span class="in">`dplyr`</span> functions for manipulating data frames.</span>
<span id="cb82-183"><a href="#cb82-183" aria-hidden="true" tabindex="-1"></a>To showcase these, we will use the <span class="in">`m_dose`</span> that we practiced reading in last chapter.</span>
<span id="cb82-184"><a href="#cb82-184" aria-hidden="true" tabindex="-1"></a>This imaginary dataset contains information on the weight lost by different strains of mice after being treated with different doses of MouseZempic®.</span>
<span id="cb82-185"><a href="#cb82-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-188"><a href="#cb82-188" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-189"><a href="#cb82-189" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb82-190"><a href="#cb82-190" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the data, like we did in chapter 1</span></span>
<span id="cb82-191"><a href="#cb82-191" aria-hidden="true" tabindex="-1"></a>m_dose <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb82-192"><a href="#cb82-192" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-193"><a href="#cb82-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-196"><a href="#cb82-196" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-197"><a href="#cb82-197" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb82-198"><a href="#cb82-198" aria-hidden="true" tabindex="-1"></a><span class="co"># just for rendering the book not for students to see</span></span>
<span id="cb82-199"><a href="#cb82-199" aria-hidden="true" tabindex="-1"></a>m_dose <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb82-200"><a href="#cb82-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-201"><a href="#cb82-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-202"><a href="#cb82-202" aria-hidden="true" tabindex="-1"></a>Before we start, let's use what we learned in the previous chapter to take a look at <span class="in">`m_dose`</span>:</span>
<span id="cb82-203"><a href="#cb82-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-206"><a href="#cb82-206" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-207"><a href="#cb82-207" aria-hidden="true" tabindex="-1"></a><span class="co"># it's a tibble, so prints nicely</span></span>
<span id="cb82-208"><a href="#cb82-208" aria-hidden="true" tabindex="-1"></a>m_dose</span>
<span id="cb82-209"><a href="#cb82-209" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-210"><a href="#cb82-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-211"><a href="#cb82-211" aria-hidden="true" tabindex="-1"></a>You might also like to use <span class="in">`View()`</span> to open the data in a separate window and get a closer look.</span>
<span id="cb82-212"><a href="#cb82-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-213"><a href="#cb82-213" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Using RStudio autocomplete"}</span>
<span id="cb82-214"><a href="#cb82-214" aria-hidden="true" tabindex="-1"></a>Although it's great to give our data a descriptive name like <span class="in">`m_dose`</span>, it can be a bit of a pain to type out every time.</span>
<span id="cb82-215"><a href="#cb82-215" aria-hidden="true" tabindex="-1"></a>Luckily, RStudio has a handy autocomplete feature that can solve this problem.</span>
<span id="cb82-216"><a href="#cb82-216" aria-hidden="true" tabindex="-1"></a>Just start typing the name of the object, and you'll see it will popup:</span>
<span id="cb82-217"><a href="#cb82-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-218"><a href="#cb82-218" aria-hidden="true" tabindex="-1"></a><span class="al">![RStudio autocomplete](images/S02F01_rstudio_autocomplete.png)</span></span>
<span id="cb82-219"><a href="#cb82-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-220"><a href="#cb82-220" aria-hidden="true" tabindex="-1"></a>You can then press {{&lt; kbd Tab &gt;}} to autocomplete it.</span>
<span id="cb82-221"><a href="#cb82-221" aria-hidden="true" tabindex="-1"></a>If there are multiple objects that start with the same letters, you can use the arrow keys to cycle through the options.</span>
<span id="cb82-222"><a href="#cb82-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-223"><a href="#cb82-223" aria-hidden="true" tabindex="-1"></a>Try using autocomplete in this chapter to save yourself some typing!</span>
<span id="cb82-224"><a href="#cb82-224" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-225"><a href="#cb82-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-226"><a href="#cb82-226" aria-hidden="true" tabindex="-1"></a><span class="fu">### Sorting data {#sec-sorting}</span></span>
<span id="cb82-227"><a href="#cb82-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-228"><a href="#cb82-228" aria-hidden="true" tabindex="-1"></a>Often, one of the first things you might want to do with a dataset is sort it.</span>
<span id="cb82-229"><a href="#cb82-229" aria-hidden="true" tabindex="-1"></a>In <span class="in">`dplyr`</span>, this is called 'arranging' and is done with the <span class="in">`arrange()`</span> function.</span>
<span id="cb82-230"><a href="#cb82-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-231"><a href="#cb82-231" aria-hidden="true" tabindex="-1"></a><span class="al">![Arrange orders rows by their values in one or more columns](images/S02F02_arrange.png)</span></span>
<span id="cb82-232"><a href="#cb82-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-233"><a href="#cb82-233" aria-hidden="true" tabindex="-1"></a>By default, <span class="in">`arrange()`</span> sorts in ascending order (smallest values first).</span>
<span id="cb82-234"><a href="#cb82-234" aria-hidden="true" tabindex="-1"></a>For example, let's sort the <span class="in">`m_dose`</span> data frame by the <span class="in">`weight_lost_g`</span> column:</span>
<span id="cb82-235"><a href="#cb82-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-238"><a href="#cb82-238" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-239"><a href="#cb82-239" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-240"><a href="#cb82-240" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(weight_lost_g)</span>
<span id="cb82-241"><a href="#cb82-241" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-242"><a href="#cb82-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-243"><a href="#cb82-243" aria-hidden="true" tabindex="-1"></a>If we compare this to when we just printed our data above, we can see that the rows are now sorted so that the mice that lost the least weight are at the top.</span>
<span id="cb82-244"><a href="#cb82-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-245"><a href="#cb82-245" aria-hidden="true" tabindex="-1"></a>Sometimes you might want to sort in descending order instead (largest values first).</span>
<span id="cb82-246"><a href="#cb82-246" aria-hidden="true" tabindex="-1"></a>You can do this by putting the <span class="in">`desc()`</span> function around your column name, inside <span class="in">`arrange()`</span>:</span>
<span id="cb82-247"><a href="#cb82-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-250"><a href="#cb82-250" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-251"><a href="#cb82-251" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-252"><a href="#cb82-252" aria-hidden="true" tabindex="-1"></a>  <span class="co"># put desc() around the column name to sort in descending order</span></span>
<span id="cb82-253"><a href="#cb82-253" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb82-254"><a href="#cb82-254" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-255"><a href="#cb82-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-256"><a href="#cb82-256" aria-hidden="true" tabindex="-1"></a>Now we can see the mice that lost the most weight are at the top.</span>
<span id="cb82-257"><a href="#cb82-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-258"><a href="#cb82-258" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Comments and pipes"}</span>
<span id="cb82-259"><a href="#cb82-259" aria-hidden="true" tabindex="-1"></a>Notice how in the previous example we have written a comment in the middle of the pipe chain.</span>
<span id="cb82-260"><a href="#cb82-260" aria-hidden="true" tabindex="-1"></a>This is a good practice to help you remember what each step is doing, especially when you have a long chain of functions, and won't cause any errors as long as you make sure that the comment is on its own line.</span>
<span id="cb82-261"><a href="#cb82-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-262"><a href="#cb82-262" aria-hidden="true" tabindex="-1"></a>You can also write comments at the end of the line, just make sure it's after the pipe operator <span class="in">`%&gt;%`</span>.</span>
<span id="cb82-263"><a href="#cb82-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-264"><a href="#cb82-264" aria-hidden="true" tabindex="-1"></a>For example, these comments are allowed:</span>
<span id="cb82-265"><a href="#cb82-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-268"><a href="#cb82-268" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-269"><a href="#cb82-269" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span> <span class="co"># a comment here is fine</span></span>
<span id="cb82-270"><a href="#cb82-270" aria-hidden="true" tabindex="-1"></a>  <span class="co"># a comment here is fine</span></span>
<span id="cb82-271"><a href="#cb82-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb82-272"><a href="#cb82-272" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-273"><a href="#cb82-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-274"><a href="#cb82-274" aria-hidden="true" tabindex="-1"></a>But this will cause an error, because the <span class="in">`#`</span> is before the pipe, so R treats it as part of the comment (notice how the <span class="in">`%&gt;%`</span> has changed colour?) and doesn't know how the two lines relate to each other.</span>
<span id="cb82-275"><a href="#cb82-275" aria-hidden="true" tabindex="-1"></a>It tries to run them separately, which for the first line is ok (it will just print <span class="in">`m_dose`</span>):</span>
<span id="cb82-276"><a href="#cb82-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-279"><a href="#cb82-279" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-280"><a href="#cb82-280" aria-hidden="true" tabindex="-1"></a><span class="co">#| error: true</span></span>
<span id="cb82-281"><a href="#cb82-281" aria-hidden="true" tabindex="-1"></a>m_dose <span class="co"># this comment will cause an error %&gt;%</span></span>
<span id="cb82-282"><a href="#cb82-282" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(weight_lost_g))</span>
<span id="cb82-283"><a href="#cb82-283" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-284"><a href="#cb82-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-285"><a href="#cb82-285" aria-hidden="true" tabindex="-1"></a>But for the second line, there is an error that R doesn't know what the <span class="in">`weight_lost_g`</span> object is.</span>
<span id="cb82-286"><a href="#cb82-286" aria-hidden="true" tabindex="-1"></a>That's because it's a column in the <span class="in">`m_dose`</span> data frame, so R only knows what it is in the context of the pipe chain containing that data frame.</span>
<span id="cb82-287"><a href="#cb82-287" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-288"><a href="#cb82-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-289"><a href="#cb82-289" aria-hidden="true" tabindex="-1"></a>You can also sort by multiple columns by passing multiple column names to <span class="in">`arrange()`</span>.</span>
<span id="cb82-290"><a href="#cb82-290" aria-hidden="true" tabindex="-1"></a>For example, to sort by the strain first and then by the amount of weight lost:</span>
<span id="cb82-291"><a href="#cb82-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-294"><a href="#cb82-294" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-295"><a href="#cb82-295" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb82-296"><a href="#cb82-296" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-297"><a href="#cb82-297" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g)</span>
<span id="cb82-298"><a href="#cb82-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-299"><a href="#cb82-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-300"><a href="#cb82-300" aria-hidden="true" tabindex="-1"></a>This will sort the data frame by strain (according to alphabetical order, as it is a character column), and within each strain, they are then sorted by the amount of weight lost.</span>
<span id="cb82-301"><a href="#cb82-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-302"><a href="#cb82-302" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Piping into View()"}</span>
<span id="cb82-303"><a href="#cb82-303" aria-hidden="true" tabindex="-1"></a>In the above example, we sorted the data by strain and then by weight lost, but because there are so many mice in each strain, the preview shown in our console doesn't allow us to see the full effect of the sorting.</span>
<span id="cb82-304"><a href="#cb82-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-305"><a href="#cb82-305" aria-hidden="true" tabindex="-1"></a>One handy trick you can use with pipes is to add <span class="in">`View()`</span> at the end of your chain to open the data in a separate window.</span>
<span id="cb82-306"><a href="#cb82-306" aria-hidden="true" tabindex="-1"></a>Try running this code, and you'll be able to scroll through the full dataset to check that the other mouse strains have also been sorted correctly:</span>
<span id="cb82-307"><a href="#cb82-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-310"><a href="#cb82-310" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-311"><a href="#cb82-311" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb82-312"><a href="#cb82-312" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by strain first, then by weight lost</span></span>
<span id="cb82-313"><a href="#cb82-313" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-314"><a href="#cb82-314" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, weight_lost_g) <span class="sc">%&gt;%</span></span>
<span id="cb82-315"><a href="#cb82-315" aria-hidden="true" tabindex="-1"></a>  <span class="fu">View</span>()</span>
<span id="cb82-316"><a href="#cb82-316" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-317"><a href="#cb82-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-318"><a href="#cb82-318" aria-hidden="true" tabindex="-1"></a>This is a great way to check that your code has actually done what you intended!</span>
<span id="cb82-319"><a href="#cb82-319" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-320"><a href="#cb82-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-321"><a href="#cb82-321" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Extracting rows with the smallest or largest values {#sec-sliceMinMax}</span></span>
<span id="cb82-322"><a href="#cb82-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-323"><a href="#cb82-323" aria-hidden="true" tabindex="-1"></a>Slice functions are used to select rows based on their position in the data frame.</span>
<span id="cb82-324"><a href="#cb82-324" aria-hidden="true" tabindex="-1"></a>The <span class="in">`slice_min()`</span> and <span class="in">`slice_max()`</span> functions are particularly useful, because they allow you to select the rows with the smallest or largest values in a particular column.</span>
<span id="cb82-325"><a href="#cb82-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-326"><a href="#cb82-326" aria-hidden="true" tabindex="-1"></a>This is equivalent to using <span class="in">`arrange()`</span> followed by <span class="in">`head()`</span>, but is more concise:</span>
<span id="cb82-327"><a href="#cb82-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-330"><a href="#cb82-330" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-331"><a href="#cb82-331" aria-hidden="true" tabindex="-1"></a><span class="co"># get the 10 mice with the lowest drug dose</span></span>
<span id="cb82-332"><a href="#cb82-332" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-333"><a href="#cb82-333" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_min() requires the column to sort by, and n = the number of rows to keep</span></span>
<span id="cb82-334"><a href="#cb82-334" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_min</span>(drug_dose_g, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb82-335"><a href="#cb82-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-336"><a href="#cb82-336" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb82-337"><a href="#cb82-337" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-338"><a href="#cb82-338" aria-hidden="true" tabindex="-1"></a>  <span class="co"># slice_max() has the same arguments as slice_min()</span></span>
<span id="cb82-339"><a href="#cb82-339" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>)</span>
<span id="cb82-340"><a href="#cb82-340" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-341"><a href="#cb82-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-342"><a href="#cb82-342" aria-hidden="true" tabindex="-1"></a>But wait— neither of those pieces of code actually gave the number of rows we asked for!</span>
<span id="cb82-343"><a href="#cb82-343" aria-hidden="true" tabindex="-1"></a>In the first example, we asked for the 10 mice with the lowest drug dose, but we got 13.</span>
<span id="cb82-344"><a href="#cb82-344" aria-hidden="true" tabindex="-1"></a>And in the second example, we asked for the top 5 mice that lost the most weight, but we got 6.</span>
<span id="cb82-345"><a href="#cb82-345" aria-hidden="true" tabindex="-1"></a>Why aren't the <span class="in">`slice_`</span> functions behaving as expected?</span>
<span id="cb82-346"><a href="#cb82-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-347"><a href="#cb82-347" aria-hidden="true" tabindex="-1"></a>If we take a look at the help page (type <span class="in">`?slice_min`</span> in the console), we learn that <span class="in">`slice_min()`</span> and <span class="in">`slice_max()`</span> have an argument called <span class="in">`with_ties`</span> that is set to <span class="in">`TRUE`</span> by default.</span>
<span id="cb82-348"><a href="#cb82-348" aria-hidden="true" tabindex="-1"></a>If we want to make sure we only get the number of rows we asked for, we would have to set it to <span class="in">`FALSE`</span>, like so:</span>
<span id="cb82-349"><a href="#cb82-349" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-352"><a href="#cb82-352" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-353"><a href="#cb82-353" aria-hidden="true" tabindex="-1"></a><span class="co"># get the top 5 mice that lost the most weight</span></span>
<span id="cb82-354"><a href="#cb82-354" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-355"><a href="#cb82-355" aria-hidden="true" tabindex="-1"></a>  <span class="co"># no ties allowed!</span></span>
<span id="cb82-356"><a href="#cb82-356" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>)</span>
<span id="cb82-357"><a href="#cb82-357" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-358"><a href="#cb82-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-359"><a href="#cb82-359" aria-hidden="true" tabindex="-1"></a>This is an important lesson: sometimes functions will behave in a way that is unexpected, and you might need to read their help page or use other guides/google/AI to understand why.</span>
<span id="cb82-360"><a href="#cb82-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-361"><a href="#cb82-361" aria-hidden="true" tabindex="-1"></a>::::::::::::::::::::: {.callout-important title="Practice exercises"}</span>
<span id="cb82-362"><a href="#cb82-362" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb82-363"><a href="#cb82-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-364"><a href="#cb82-364" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-365"><a href="#cb82-365" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span></span>
<span id="cb82-366"><a href="#cb82-366" aria-hidden="true" tabindex="-1"></a>Which code would you use to sort the <span class="in">`m_dose`</span> data frame from biggest to smallest initial weight?</span>
<span id="cb82-367"><a href="#cb82-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-368"><a href="#cb82-368" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-369"><a href="#cb82-369" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-370"><a href="#cb82-370" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% sort(initial_weight_g)`</span></span>
<span id="cb82-371"><a href="#cb82-371" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-372"><a href="#cb82-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-373"><a href="#cb82-373" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-374"><a href="#cb82-374" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% arrange(initial_weight_g)`</span></span>
<span id="cb82-375"><a href="#cb82-375" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-376"><a href="#cb82-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-377"><a href="#cb82-377" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-378"><a href="#cb82-378" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% sort(descending(initial_weight_g))`</span></span>
<span id="cb82-379"><a href="#cb82-379" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-380"><a href="#cb82-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-381"><a href="#cb82-381" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-382"><a href="#cb82-382" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% arrange(desc(initial_weight_g))`</span></span>
<span id="cb82-383"><a href="#cb82-383" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-384"><a href="#cb82-384" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-385"><a href="#cb82-385" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-386"><a href="#cb82-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-387"><a href="#cb82-387" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-388"><a href="#cb82-388" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span></span>
<span id="cb82-389"><a href="#cb82-389" aria-hidden="true" tabindex="-1"></a>Which code would you use to extract the 3 mice with the highest initial weight from the <span class="in">`m_dose`</span> data frame?</span>
<span id="cb82-390"><a href="#cb82-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-391"><a href="#cb82-391" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-392"><a href="#cb82-392" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-393"><a href="#cb82-393" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% slice_max(initial_weight_g, n = 3)`</span></span>
<span id="cb82-394"><a href="#cb82-394" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-395"><a href="#cb82-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-396"><a href="#cb82-396" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-397"><a href="#cb82-397" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% arrange(desc(initial_weight_g))`</span></span>
<span id="cb82-398"><a href="#cb82-398" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-399"><a href="#cb82-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-400"><a href="#cb82-400" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-401"><a href="#cb82-401" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% slice_min(initial_weight_g, n = 3)`</span></span>
<span id="cb82-402"><a href="#cb82-402" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-403"><a href="#cb82-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-404"><a href="#cb82-404" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-405"><a href="#cb82-405" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% arrange(initial_weight_g)`</span></span>
<span id="cb82-406"><a href="#cb82-406" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-407"><a href="#cb82-407" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-408"><a href="#cb82-408" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-409"><a href="#cb82-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-410"><a href="#cb82-410" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-411"><a href="#cb82-411" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span></span>
<span id="cb82-412"><a href="#cb82-412" aria-hidden="true" tabindex="-1"></a>I've written the below code, but one of the comments is messing it up!</span>
<span id="cb82-413"><a href="#cb82-413" aria-hidden="true" tabindex="-1"></a>Which one?</span>
<span id="cb82-414"><a href="#cb82-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-417"><a href="#cb82-417" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-418"><a href="#cb82-418" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb82-419"><a href="#cb82-419" aria-hidden="true" tabindex="-1"></a><span class="co"># comment A</span></span>
<span id="cb82-420"><a href="#cb82-420" aria-hidden="true" tabindex="-1"></a>m_dose <span class="co"># comment B %&gt;%</span></span>
<span id="cb82-421"><a href="#cb82-421" aria-hidden="true" tabindex="-1"></a>  <span class="co"># comment C</span></span>
<span id="cb82-422"><a href="#cb82-422" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(weight_lost_g, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">with_ties =</span> <span class="cn">FALSE</span>) <span class="co"># comment D</span></span>
<span id="cb82-423"><a href="#cb82-423" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-424"><a href="#cb82-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-425"><a href="#cb82-425" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-426"><a href="#cb82-426" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-427"><a href="#cb82-427" aria-hidden="true" tabindex="-1"></a>Comment A</span>
<span id="cb82-428"><a href="#cb82-428" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-429"><a href="#cb82-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-430"><a href="#cb82-430" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-431"><a href="#cb82-431" aria-hidden="true" tabindex="-1"></a>Comment B</span>
<span id="cb82-432"><a href="#cb82-432" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-433"><a href="#cb82-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-434"><a href="#cb82-434" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-435"><a href="#cb82-435" aria-hidden="true" tabindex="-1"></a>Comment C</span>
<span id="cb82-436"><a href="#cb82-436" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-437"><a href="#cb82-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-438"><a href="#cb82-438" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-439"><a href="#cb82-439" aria-hidden="true" tabindex="-1"></a>Comment D</span>
<span id="cb82-440"><a href="#cb82-440" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-441"><a href="#cb82-441" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-442"><a href="#cb82-442" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-443"><a href="#cb82-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-444"><a href="#cb82-444" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb82-445"><a href="#cb82-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-446"><a href="#cb82-446" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb82-447"><a href="#cb82-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-448"><a href="#cb82-448" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct code to sort the <span class="in">`m_dose`</span> data frame from biggest to smallest initial weight is <span class="in">`m_dose %&gt;% arrange(desc(initial_weight_g))`</span>.</span>
<span id="cb82-449"><a href="#cb82-449" aria-hidden="true" tabindex="-1"></a>    The <span class="in">`arrange()`</span> function is used to sort the data frame (although there is a <span class="in">`sort()`</span> function in R, that's not part of dplyr and won't work the same way), and the <span class="in">`desc()`</span> function is used to sort in descending order.</span>
<span id="cb82-450"><a href="#cb82-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-451"><a href="#cb82-451" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct code to extract the 3 mice with the highest initial weight from the <span class="in">`m_dose`</span> data frame is <span class="in">`m_dose %&gt;% slice_max(initial_weight_g, n = 3)`</span>.</span>
<span id="cb82-452"><a href="#cb82-452" aria-hidden="true" tabindex="-1"></a>    The <span class="in">`slice_max()`</span> function is used to select the rows with the largest values in the <span class="in">`initial_weight_g`</span> column, and the <span class="in">`n = 3`</span> argument specifies that we want to keep 3 rows.</span>
<span id="cb82-453"><a href="#cb82-453" aria-hidden="true" tabindex="-1"></a>    The <span class="in">`arrange()`</span> function is not needed in this case, because <span class="in">`slice_max()`</span> will automatically sort the data frame by the specified column.</span>
<span id="cb82-454"><a href="#cb82-454" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-455"><a href="#cb82-455" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The comment that is messing up the code is Comment B. The <span class="in">`#`</span> symbol is before the pipe operator <span class="in">`%&gt;%`</span>, so R treats it as part of the comment and this breaks our chain of pipes.</span>
<span id="cb82-456"><a href="#cb82-456" aria-hidden="true" tabindex="-1"></a>    The other comments are fine, because they are either at the end of the line or on their own line.</span>
<span id="cb82-457"><a href="#cb82-457" aria-hidden="true" tabindex="-1"></a>    Basically, if a comment is changing the colour of the pipe operator (or any other bits of your code), it's in the wrong place!</span>
<span id="cb82-458"><a href="#cb82-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-459"><a href="#cb82-459" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb82-460"><a href="#cb82-460" aria-hidden="true" tabindex="-1"></a>:::::::::::::::::::::</span>
<span id="cb82-461"><a href="#cb82-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-462"><a href="#cb82-462" aria-hidden="true" tabindex="-1"></a><span class="fu">### Filtering data (rows) {#sec-filter}</span></span>
<span id="cb82-463"><a href="#cb82-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-464"><a href="#cb82-464" aria-hidden="true" tabindex="-1"></a><span class="al">![Filter allows you to filter rows using a logical test](images/S02F03_filter.png)</span></span>
<span id="cb82-465"><a href="#cb82-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-466"><a href="#cb82-466" aria-hidden="true" tabindex="-1"></a>In <span class="in">`dplyr`</span>, the <span class="in">`filter()`</span> function is used to subset rows based on their values.</span>
<span id="cb82-467"><a href="#cb82-467" aria-hidden="true" tabindex="-1"></a>You provide a logical test, and <span class="in">`filter()`</span> will keep the rows where the test is <span class="in">`TRUE`</span>.</span>
<span id="cb82-468"><a href="#cb82-468" aria-hidden="true" tabindex="-1"></a>We can write these tests using the comparison operators we learned in the previous chapter (e.g. <span class="in">`==`</span>, <span class="in">`&lt;`</span> and <span class="in">`!=`</span>, see <span class="co">[</span><span class="ot">Section @sec-comparisons</span><span class="co">]</span>).</span>
<span id="cb82-469"><a href="#cb82-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-470"><a href="#cb82-470" aria-hidden="true" tabindex="-1"></a>For example, to filter the <span class="in">`m_dose`</span> data frame to only include mice that lost more than 6g:</span>
<span id="cb82-471"><a href="#cb82-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-474"><a href="#cb82-474" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-475"><a href="#cb82-475" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-476"><a href="#cb82-476" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">6</span>)</span>
<span id="cb82-477"><a href="#cb82-477" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-478"><a href="#cb82-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-479"><a href="#cb82-479" aria-hidden="true" tabindex="-1"></a>Or to only include mice from cage 3E:</span>
<span id="cb82-480"><a href="#cb82-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-483"><a href="#cb82-483" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-484"><a href="#cb82-484" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-485"><a href="#cb82-485" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember that == is used for testing equality</span></span>
<span id="cb82-486"><a href="#cb82-486" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span>) <span class="co"># don't forget the quotes either!</span></span>
<span id="cb82-487"><a href="#cb82-487" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-488"><a href="#cb82-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-489"><a href="#cb82-489" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Combining logical tests</span></span>
<span id="cb82-490"><a href="#cb82-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-491"><a href="#cb82-491" aria-hidden="true" tabindex="-1"></a>Sometimes we want to filter based on multiple conditions.</span>
<span id="cb82-492"><a href="#cb82-492" aria-hidden="true" tabindex="-1"></a>Here we will show some more advanced operators that can be used to combine logical tests.</span>
<span id="cb82-493"><a href="#cb82-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-494"><a href="#cb82-494" aria-hidden="true" tabindex="-1"></a>The <span class="in">`&amp;`</span> operator is used to combine two logical tests with an 'and' condition.</span>
<span id="cb82-495"><a href="#cb82-495" aria-hidden="true" tabindex="-1"></a>For example, to filter the data frame to only include mice that have a tail length greater than 19mm and are female:</span>
<span id="cb82-496"><a href="#cb82-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-499"><a href="#cb82-499" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-500"><a href="#cb82-500" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-501"><a href="#cb82-501" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(tail_length_mm <span class="sc">&gt;</span> <span class="dv">19</span> <span class="sc">&amp;</span> sex <span class="sc">==</span> <span class="st">"female"</span>)</span>
<span id="cb82-502"><a href="#cb82-502" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-503"><a href="#cb82-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-504"><a href="#cb82-504" aria-hidden="true" tabindex="-1"></a>The <span class="in">`|`</span> operator is used to combine two logical tests with an 'or' condition.</span>
<span id="cb82-505"><a href="#cb82-505" aria-hidden="true" tabindex="-1"></a>For example, to filter the data frame to only include mice that have an initial weight less than 35g or a tail length less than 14mm:</span>
<span id="cb82-506"><a href="#cb82-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-509"><a href="#cb82-509" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-510"><a href="#cb82-510" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-511"><a href="#cb82-511" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(initial_weight_g <span class="sc">&lt;</span> <span class="dv">35</span> <span class="sc">|</span> tail_length_mm <span class="sc">&lt;</span> <span class="dv">14</span>)</span>
<span id="cb82-512"><a href="#cb82-512" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-513"><a href="#cb82-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-514"><a href="#cb82-514" aria-hidden="true" tabindex="-1"></a>The <span class="in">`%in%`</span> operator can be used to filter based on a vector of multiple values (<span class="in">`c(x, y)`</span>).</span>
<span id="cb82-515"><a href="#cb82-515" aria-hidden="true" tabindex="-1"></a>It's particularly useful when you have a few character values you want to filter on, as it is shorter to type than <span class="in">`|`</span> (or).</span>
<span id="cb82-516"><a href="#cb82-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-517"><a href="#cb82-517" aria-hidden="true" tabindex="-1"></a>For example, to filter the data frame to only include mice from cages 3E or 1A, we could use <span class="in">`|`</span> like this:</span>
<span id="cb82-518"><a href="#cb82-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-521"><a href="#cb82-521" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-522"><a href="#cb82-522" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-523"><a href="#cb82-523" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">==</span> <span class="st">"3E"</span> <span class="sc">|</span> cage_number <span class="sc">==</span> <span class="st">"1A"</span>)</span>
<span id="cb82-524"><a href="#cb82-524" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-525"><a href="#cb82-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-526"><a href="#cb82-526" aria-hidden="true" tabindex="-1"></a>Or we could use <span class="in">`%in%`</span> like this:</span>
<span id="cb82-527"><a href="#cb82-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-530"><a href="#cb82-530" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-531"><a href="#cb82-531" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-532"><a href="#cb82-532" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cage_number <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"3E"</span>, <span class="st">"1A"</span>))</span>
<span id="cb82-533"><a href="#cb82-533" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-534"><a href="#cb82-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-535"><a href="#cb82-535" aria-hidden="true" tabindex="-1"></a>::::::::::::::::::::: {.callout-important title="Practice exercises"}</span>
<span id="cb82-536"><a href="#cb82-536" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb82-537"><a href="#cb82-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-538"><a href="#cb82-538" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-539"><a href="#cb82-539" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span></span>
<span id="cb82-540"><a href="#cb82-540" aria-hidden="true" tabindex="-1"></a>Which code would you use to filter the <span class="in">`m_dose`</span> data frame to only include mice from replicate 2?</span>
<span id="cb82-541"><a href="#cb82-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-542"><a href="#cb82-542" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-543"><a href="#cb82-543" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-544"><a href="#cb82-544" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(replicate == 2)`</span></span>
<span id="cb82-545"><a href="#cb82-545" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-546"><a href="#cb82-546" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-547"><a href="#cb82-547" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-548"><a href="#cb82-548" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(replicate == rep2)`</span></span>
<span id="cb82-549"><a href="#cb82-549" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-550"><a href="#cb82-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-551"><a href="#cb82-551" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-552"><a href="#cb82-552" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(replicate == "rep2")`</span></span>
<span id="cb82-553"><a href="#cb82-553" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-554"><a href="#cb82-554" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-555"><a href="#cb82-555" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-556"><a href="#cb82-556" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(replicate = "rep2")`</span></span>
<span id="cb82-557"><a href="#cb82-557" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-558"><a href="#cb82-558" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-559"><a href="#cb82-559" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-560"><a href="#cb82-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-561"><a href="#cb82-561" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-562"><a href="#cb82-562" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span></span>
<span id="cb82-563"><a href="#cb82-563" aria-hidden="true" tabindex="-1"></a>What is NOT a valid way to filter the <span class="in">`m_dose`</span> data frame to only include mice that lost more than 4g, and have an initial weight less than 40g?</span>
<span id="cb82-564"><a href="#cb82-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-565"><a href="#cb82-565" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-566"><a href="#cb82-566" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-567"><a href="#cb82-567" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(weight_lost_g &gt; 4) %&gt;% filter(initial_weight_g &lt; 40)`</span></span>
<span id="cb82-568"><a href="#cb82-568" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-569"><a href="#cb82-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-570"><a href="#cb82-570" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-571"><a href="#cb82-571" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)`</span></span>
<span id="cb82-572"><a href="#cb82-572" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-573"><a href="#cb82-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-574"><a href="#cb82-574" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-575"><a href="#cb82-575" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(weight_lost_g &gt; 4 &amp; initial_weight_g &lt; 40)`</span></span>
<span id="cb82-576"><a href="#cb82-576" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-577"><a href="#cb82-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-578"><a href="#cb82-578" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-579"><a href="#cb82-579" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(initial_weight_g &lt; 40) %&gt;% filter(weight_lost_g &gt; 4)`</span></span>
<span id="cb82-580"><a href="#cb82-580" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-581"><a href="#cb82-581" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-582"><a href="#cb82-582" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-583"><a href="#cb82-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-584"><a href="#cb82-584" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-585"><a href="#cb82-585" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span></span>
<span id="cb82-586"><a href="#cb82-586" aria-hidden="true" tabindex="-1"></a>Which option correctly describes what the following code is doing?</span>
<span id="cb82-587"><a href="#cb82-587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-590"><a href="#cb82-590" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-591"><a href="#cb82-591" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb82-592"><a href="#cb82-592" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-593"><a href="#cb82-593" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(mouse_strain <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"BALB C"</span>, <span class="st">"Black 6"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb82-594"><a href="#cb82-594" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;</span> <span class="dv">3</span> <span class="sc">&amp;</span> weight_lost_g <span class="sc">&lt;</span> <span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb82-595"><a href="#cb82-595" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(drug_dose_g))</span>
<span id="cb82-596"><a href="#cb82-596" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-597"><a href="#cb82-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-598"><a href="#cb82-598" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-599"><a href="#cb82-599" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-600"><a href="#cb82-600" aria-hidden="true" tabindex="-1"></a>Filters the data frame to remove mice from the "BALB C" and "Black 6" strains, who only lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</span>
<span id="cb82-601"><a href="#cb82-601" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-602"><a href="#cb82-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-603"><a href="#cb82-603" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-604"><a href="#cb82-604" aria-hidden="true" tabindex="-1"></a>Filters the data frame to remove mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</span>
<span id="cb82-605"><a href="#cb82-605" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-606"><a href="#cb82-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-607"><a href="#cb82-607" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-608"><a href="#cb82-608" aria-hidden="true" tabindex="-1"></a>Filters the data frame to only include mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose.</span>
<span id="cb82-609"><a href="#cb82-609" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-610"><a href="#cb82-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-611"><a href="#cb82-611" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-612"><a href="#cb82-612" aria-hidden="true" tabindex="-1"></a>Filters the data frame to only include mice from the "BALB C" and "Black 6" strains, that lost between 3 and 5g of weight, and then sorts the data frame by drug dose in descending order.</span>
<span id="cb82-613"><a href="#cb82-613" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-614"><a href="#cb82-614" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-615"><a href="#cb82-615" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-616"><a href="#cb82-616" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-617"><a href="#cb82-617" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb82-618"><a href="#cb82-618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-619"><a href="#cb82-619" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb82-620"><a href="#cb82-620" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-621"><a href="#cb82-621" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The correct code to filter the <span class="in">`m_dose`</span> data frame to only include mice from replicate 2 is <span class="in">`m_dose %&gt;% filter(replicate == "rep2")`</span>.</span>
<span id="cb82-622"><a href="#cb82-622" aria-hidden="true" tabindex="-1"></a>    Option A is incorrect because <span class="in">`2`</span> is not a value of <span class="in">`replicate`</span> (when filtering you need to know what values are actually in your columns! So make sure to <span class="in">`View()`</span> your data first).</span>
<span id="cb82-623"><a href="#cb82-623" aria-hidden="true" tabindex="-1"></a>    Option B is incorrect because the replicate column is a character column, so you need to use quotes around the value you are filtering on.</span>
<span id="cb82-624"><a href="#cb82-624" aria-hidden="true" tabindex="-1"></a>    Option D is incorrect because <span class="in">`=`</span> is not the correct way to test for equality, you need to use <span class="in">`==`</span>.</span>
<span id="cb82-625"><a href="#cb82-625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-626"><a href="#cb82-626" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The invalid option is <span class="in">`m_dose %&gt;% filter(weight_lost_g &gt; 4) %&gt;% (initial_weight_g &lt; 40)`</span>.</span>
<span id="cb82-627"><a href="#cb82-627" aria-hidden="true" tabindex="-1"></a>    This is because the second filtering step is missing the name of the filter function, so R doesn't know what to do with <span class="in">`(initial_weight_g &lt; 40)`</span>.</span>
<span id="cb82-628"><a href="#cb82-628" aria-hidden="true" tabindex="-1"></a>    The other options are valid ways to filter the data frame based on the specified conditions; note that we can use multiple <span class="in">`filter()`</span> functions in a row to apply multiple conditions, or the <span class="in">`&amp;`</span> operator to combine them into a single <span class="in">`filter()`</span> function.</span>
<span id="cb82-629"><a href="#cb82-629" aria-hidden="true" tabindex="-1"></a>    It's just a matter of personal preference.</span>
<span id="cb82-630"><a href="#cb82-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-631"><a href="#cb82-631" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct description of the code is that it filters the data frame to only include mice from the "BALB C" and "Black 6" strains, then filters those further to only those that lost between 3 and 5g of weight, and finally sorts the data frame by drug dose in descending order.</span>
<span id="cb82-632"><a href="#cb82-632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-633"><a href="#cb82-633" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb82-634"><a href="#cb82-634" aria-hidden="true" tabindex="-1"></a>:::::::::::::::::::::</span>
<span id="cb82-635"><a href="#cb82-635" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-636"><a href="#cb82-636" aria-hidden="true" tabindex="-1"></a><span class="fu">### Dealing with missing values {#sec-missing}</span></span>
<span id="cb82-637"><a href="#cb82-637" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-638"><a href="#cb82-638" aria-hidden="true" tabindex="-1"></a>Missing values are a common problem in real-world datasets.</span>
<span id="cb82-639"><a href="#cb82-639" aria-hidden="true" tabindex="-1"></a>In R, missing values are represented by <span class="in">`NA`</span>.</span>
<span id="cb82-640"><a href="#cb82-640" aria-hidden="true" tabindex="-1"></a>In fact, if you look at the <span class="in">`m_dose`</span> data frame we've been using, you'll see that some of the cells contain <span class="in">`NA`</span>: try spotting them with the <span class="in">`View()`</span> function.</span>
<span id="cb82-641"><a href="#cb82-641" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-642"><a href="#cb82-642" aria-hidden="true" tabindex="-1"></a>You can also find missing values in a data frame using the <span class="in">`is.na()`</span> function in combination with <span class="in">`filter()`</span>.</span>
<span id="cb82-643"><a href="#cb82-643" aria-hidden="true" tabindex="-1"></a>For example, to find all the rows in the <span class="in">`m_dose`</span> data frame that have a missing value for the <span class="in">`drug_dose_g`</span> column:</span>
<span id="cb82-644"><a href="#cb82-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-647"><a href="#cb82-647" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-648"><a href="#cb82-648" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-649"><a href="#cb82-649" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(drug_dose_g))</span>
<span id="cb82-650"><a href="#cb82-650" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-651"><a href="#cb82-651" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-652"><a href="#cb82-652" aria-hidden="true" tabindex="-1"></a>The problem with missing values is that they can cause problems when you try to perform calculations on your data.</span>
<span id="cb82-653"><a href="#cb82-653" aria-hidden="true" tabindex="-1"></a>For example, if you try to calculate the mean of a column that contains even a single missing value, the result will also be <span class="in">`NA`</span>:</span>
<span id="cb82-654"><a href="#cb82-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-657"><a href="#cb82-657" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-658"><a href="#cb82-658" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb82-659"><a href="#cb82-659" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from chapter 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb82-660"><a href="#cb82-660" aria-hidden="true" tabindex="-1"></a>m_dose<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span>
<span id="cb82-661"><a href="#cb82-661" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-662"><a href="#cb82-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-663"><a href="#cb82-663" aria-hidden="true" tabindex="-1"></a><span class="in">`NA`</span> values in R are therefore referred to as 'contagious': if you put an <span class="in">`NA`</span> in you usually get an <span class="in">`NA`</span> out.</span>
<span id="cb82-664"><a href="#cb82-664" aria-hidden="true" tabindex="-1"></a>If you think about it, that makes sense— when we don't know the value of a particular mouse's drug dose, how can we calculate the average?</span>
<span id="cb82-665"><a href="#cb82-665" aria-hidden="true" tabindex="-1"></a>That missing value could be anything.</span>
<span id="cb82-666"><a href="#cb82-666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-667"><a href="#cb82-667" aria-hidden="true" tabindex="-1"></a>For this reason, it's important to deal with missing values before performing calculations.</span>
<span id="cb82-668"><a href="#cb82-668" aria-hidden="true" tabindex="-1"></a>Many functions in R will have an argument called <span class="in">`na.rm`</span> that you can set to <span class="in">`TRUE`</span> to remove missing values before performing the calculation.</span>
<span id="cb82-669"><a href="#cb82-669" aria-hidden="true" tabindex="-1"></a>For example, to calculate the mean of the <span class="in">`drug_dose_g`</span> column with the missing values excluded:</span>
<span id="cb82-670"><a href="#cb82-670" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-673"><a href="#cb82-673" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-674"><a href="#cb82-674" aria-hidden="true" tabindex="-1"></a><span class="co"># try to calculate the mean of the drug_dose_g column</span></span>
<span id="cb82-675"><a href="#cb82-675" aria-hidden="true" tabindex="-1"></a><span class="co"># remember from chapter 1 that we can use $ to access columns in a data frame</span></span>
<span id="cb82-676"><a href="#cb82-676" aria-hidden="true" tabindex="-1"></a>m_dose<span class="sc">$</span>drug_dose_g <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-677"><a href="#cb82-677" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-678"><a href="#cb82-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-679"><a href="#cb82-679" aria-hidden="true" tabindex="-1"></a>This time, the result is a number, because the missing values have been removed before the calculation.</span>
<span id="cb82-680"><a href="#cb82-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-681"><a href="#cb82-681" aria-hidden="true" tabindex="-1"></a>But not all functions have an <span class="in">`na.rm`</span> argument.</span>
<span id="cb82-682"><a href="#cb82-682" aria-hidden="true" tabindex="-1"></a>In these cases, you can remove rows with missing values.</span>
<span id="cb82-683"><a href="#cb82-683" aria-hidden="true" tabindex="-1"></a>This can be done for a single column, using the <span class="in">`filter()`</span> function together with <span class="in">`is.na()`</span>:</span>
<span id="cb82-684"><a href="#cb82-684" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-687"><a href="#cb82-687" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-688"><a href="#cb82-688" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in the drug_dose_g column</span></span>
<span id="cb82-689"><a href="#cb82-689" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-690"><a href="#cb82-690" aria-hidden="true" tabindex="-1"></a>  <span class="co"># remember the ! means 'not', it negates the result of is.na()</span></span>
<span id="cb82-691"><a href="#cb82-691" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(drug_dose_g))</span>
<span id="cb82-692"><a href="#cb82-692" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-693"><a href="#cb82-693" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-694"><a href="#cb82-694" aria-hidden="true" tabindex="-1"></a>Or, you can remove rows with missing values in any column using the <span class="in">`na.omit()`</span> or <span class="in">`drop_na()`</span> function:</span>
<span id="cb82-695"><a href="#cb82-695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-698"><a href="#cb82-698" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-699"><a href="#cb82-699" aria-hidden="true" tabindex="-1"></a><span class="co"># remove rows with missing values in any column</span></span>
<span id="cb82-700"><a href="#cb82-700" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-701"><a href="#cb82-701" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb82-702"><a href="#cb82-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-703"><a href="#cb82-703" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-704"><a href="#cb82-704" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>()</span>
<span id="cb82-705"><a href="#cb82-705" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-706"><a href="#cb82-706" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-707"><a href="#cb82-707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-708"><a href="#cb82-708" aria-hidden="true" tabindex="-1"></a>Sometimes, instead of removing rows with missing values, you might want to replace them with a specific value.</span>
<span id="cb82-709"><a href="#cb82-709" aria-hidden="true" tabindex="-1"></a>This can be done using the <span class="in">`replace_na()`</span> function from the <span class="in">`tidyr`</span> package.</span>
<span id="cb82-710"><a href="#cb82-710" aria-hidden="true" tabindex="-1"></a><span class="in">`replace_na()`</span> takes a <span class="in">`list()`</span> which contains each of the column names you want to edit, and the value that should be used.</span>
<span id="cb82-711"><a href="#cb82-711" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-712"><a href="#cb82-712" aria-hidden="true" tabindex="-1"></a>For example, to replace missing values in the <span class="in">`weight_lost_g`</span> columns with 0, replace missing values in the <span class="in">`sex`</span> column with 'unknown' and leave the rest of the data frame unchanged:</span>
<span id="cb82-713"><a href="#cb82-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-716"><a href="#cb82-716" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-717"><a href="#cb82-717" aria-hidden="true" tabindex="-1"></a><span class="co"># replace missing values in the drug_dose_g column with 0</span></span>
<span id="cb82-718"><a href="#cb82-718" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-719"><a href="#cb82-719" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we need to provide the column_names = values_to_replace</span></span>
<span id="cb82-720"><a href="#cb82-720" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this needs to be contained within a list()</span></span>
<span id="cb82-721"><a href="#cb82-721" aria-hidden="true" tabindex="-1"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">weight_lost_g =</span> <span class="dv">0</span>, <span class="at">sex =</span> <span class="st">"unknown"</span>))</span>
<span id="cb82-722"><a href="#cb82-722" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-723"><a href="#cb82-723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-724"><a href="#cb82-724" aria-hidden="true" tabindex="-1"></a>When deciding how to handle missing values, you might have prior knowledge that <span class="in">`NA`</span> should be replaced with a specific value, or you might decide that removing rows with <span class="in">`NA`</span> is the best approach for your analysis.</span>
<span id="cb82-725"><a href="#cb82-725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-726"><a href="#cb82-726" aria-hidden="true" tabindex="-1"></a>For example, maybe we knew that the mice were given a <span class="in">`weight_lost_g`</span> of <span class="in">`NA`</span> if they didn't lose any weight, it would then make sense to replace those with 0 (as we did in the code above).</span>
<span id="cb82-727"><a href="#cb82-727" aria-hidden="true" tabindex="-1"></a>However, if the <span class="in">`drug_dose_g`</span> column was missing simply because the data was lost, we might choose to remove those rows entirely.</span>
<span id="cb82-728"><a href="#cb82-728" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-729"><a href="#cb82-729" aria-hidden="true" tabindex="-1"></a>It's important to think carefully about how missing values should be handled in your analysis.</span>
<span id="cb82-730"><a href="#cb82-730" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-731"><a href="#cb82-731" aria-hidden="true" tabindex="-1"></a>::::::::::::::::::::: {.callout-important title="Practice exercises"}</span>
<span id="cb82-732"><a href="#cb82-732" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb82-733"><a href="#cb82-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-734"><a href="#cb82-734" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-735"><a href="#cb82-735" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span></span>
<span id="cb82-736"><a href="#cb82-736" aria-hidden="true" tabindex="-1"></a>What would be the result of running this R code: <span class="in">`mean(c(1, 2, 4, NA))`</span></span>
<span id="cb82-737"><a href="#cb82-737" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-738"><a href="#cb82-738" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-739"><a href="#cb82-739" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-740"><a href="#cb82-740" aria-hidden="true" tabindex="-1"></a>2.333333</span>
<span id="cb82-741"><a href="#cb82-741" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-742"><a href="#cb82-742" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-743"><a href="#cb82-743" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-744"><a href="#cb82-744" aria-hidden="true" tabindex="-1"></a>0</span>
<span id="cb82-745"><a href="#cb82-745" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-746"><a href="#cb82-746" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-747"><a href="#cb82-747" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-748"><a href="#cb82-748" aria-hidden="true" tabindex="-1"></a><span class="in">`NA`</span></span>
<span id="cb82-749"><a href="#cb82-749" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-750"><a href="#cb82-750" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-751"><a href="#cb82-751" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-752"><a href="#cb82-752" aria-hidden="true" tabindex="-1"></a>An error</span>
<span id="cb82-753"><a href="#cb82-753" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-754"><a href="#cb82-754" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-755"><a href="#cb82-755" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-756"><a href="#cb82-756" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-757"><a href="#cb82-757" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-758"><a href="#cb82-758" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span></span>
<span id="cb82-759"><a href="#cb82-759" aria-hidden="true" tabindex="-1"></a>Which line of code would you use to filter the <span class="in">`m_dose`</span> data frame to remove mice that have a missing value in the <span class="in">`tail_length_mm`</span> column?</span>
<span id="cb82-760"><a href="#cb82-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-761"><a href="#cb82-761" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-762"><a href="#cb82-762" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-763"><a href="#cb82-763" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(tail_length_mm != NA)`</span></span>
<span id="cb82-764"><a href="#cb82-764" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-765"><a href="#cb82-765" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-766"><a href="#cb82-766" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-767"><a href="#cb82-767" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(is.na(tail_length_mm))`</span></span>
<span id="cb82-768"><a href="#cb82-768" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-769"><a href="#cb82-769" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-770"><a href="#cb82-770" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-771"><a href="#cb82-771" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% na.omit()`</span></span>
<span id="cb82-772"><a href="#cb82-772" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-773"><a href="#cb82-773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-774"><a href="#cb82-774" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-775"><a href="#cb82-775" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(!is.na(tail_length_mm))`</span></span>
<span id="cb82-776"><a href="#cb82-776" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-777"><a href="#cb82-777" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-778"><a href="#cb82-778" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-779"><a href="#cb82-779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-780"><a href="#cb82-780" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-781"><a href="#cb82-781" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span></span>
<span id="cb82-782"><a href="#cb82-782" aria-hidden="true" tabindex="-1"></a>How would you replace missing values in the <span class="in">`initial_weight_g`</span> column with the value 35?</span>
<span id="cb82-783"><a href="#cb82-783" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-784"><a href="#cb82-784" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-785"><a href="#cb82-785" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-786"><a href="#cb82-786" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% replace_na(list(initial_weight_g = 35))`</span></span>
<span id="cb82-787"><a href="#cb82-787" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-788"><a href="#cb82-788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-789"><a href="#cb82-789" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-790"><a href="#cb82-790" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% replace_na(initial_weight_g = 35)`</span></span>
<span id="cb82-791"><a href="#cb82-791" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-792"><a href="#cb82-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-793"><a href="#cb82-793" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-794"><a href="#cb82-794" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% replace_na(list(initial_weight_g == 35))`</span></span>
<span id="cb82-795"><a href="#cb82-795" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-796"><a href="#cb82-796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-797"><a href="#cb82-797" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-798"><a href="#cb82-798" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% replace_na(35)`</span></span>
<span id="cb82-799"><a href="#cb82-799" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-800"><a href="#cb82-800" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-801"><a href="#cb82-801" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-802"><a href="#cb82-802" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-803"><a href="#cb82-803" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb82-804"><a href="#cb82-804" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-805"><a href="#cb82-805" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb82-806"><a href="#cb82-806" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-807"><a href="#cb82-807" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb82-808"><a href="#cb82-808" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-809"><a href="#cb82-809" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The result of running the code <span class="in">`mean(c(1, 2, 4, NA))`</span> is <span class="in">`NA`</span>. This is because the <span class="in">`NA`</span> value is 'contagious', so when you try to calculate the mean of a vector that contains an <span class="in">`NA`</span>, the result will also be <span class="in">`NA`</span>. If we wanted to calculate the mean of the vector without the <span class="in">`NA`</span>, we would need to use the <span class="in">`na.rm = TRUE`</span> argument.</span>
<span id="cb82-810"><a href="#cb82-810" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct line of code to filter the <span class="in">`m_dose`</span> data frame to remove mice that have a missing value in the <span class="in">`tail_length_mm`</span> column is <span class="in">`m_dose %&gt;% filter(!is.na(tail_length_mm))`</span>. The <span class="in">`!`</span> symbol is used to negate the result of <span class="in">`is.na()`</span>, so we are filtering to keep the rows where <span class="in">`tail_length_mm`</span> is not <span class="in">`NA`</span>. We can't use the first option with the <span class="in">`!= NA`</span> because <span class="in">`NA`</span> is a special value in R that represents missing data, and it can't be compared to anything, and the third option is incorrect because <span class="in">`na.omit()`</span> removes entire rows with missing values, rather than just filtering based on a single column.</span>
<span id="cb82-811"><a href="#cb82-811" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct line of code to replace missing values in the <span class="in">`initial_weight_g`</span> column with the value 35 is <span class="in">`m_dose %&gt;% replace_na(list(initial_weight_g = 35))`</span>. The <span class="in">`replace_na()`</span> function takes a <span class="in">`list()`</span> that contains the column names you want to replace and the values you want to replace them with. We only need to use a single equal sign here as we're not testing for equality, we're assigning a value.</span>
<span id="cb82-812"><a href="#cb82-812" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-813"><a href="#cb82-813" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb82-814"><a href="#cb82-814" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-815"><a href="#cb82-815" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb82-816"><a href="#cb82-816" aria-hidden="true" tabindex="-1"></a>:::::::::::::::::::::</span>
<span id="cb82-817"><a href="#cb82-817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-818"><a href="#cb82-818" aria-hidden="true" tabindex="-1"></a><span class="fu">### Selecting columns {#sec-select}</span></span>
<span id="cb82-819"><a href="#cb82-819" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-820"><a href="#cb82-820" aria-hidden="true" tabindex="-1"></a><span class="al">![Select allows you to select only certain columns](images/S02F04_select.png)</span></span>
<span id="cb82-821"><a href="#cb82-821" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-822"><a href="#cb82-822" aria-hidden="true" tabindex="-1"></a>While <span class="in">`filter()`</span> is used to subset rows, <span class="in">`select()`</span> is used to subset columns.</span>
<span id="cb82-823"><a href="#cb82-823" aria-hidden="true" tabindex="-1"></a>You can use <span class="in">`select()`</span> to keep only the columns you're interested in, or to drop columns you don't need.</span>
<span id="cb82-824"><a href="#cb82-824" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-825"><a href="#cb82-825" aria-hidden="true" tabindex="-1"></a>The <span class="in">`select()`</span> function takes the names of the columns that you want to keep/remove (no vector notation <span class="in">`c()`</span> or quotation marks <span class="in">`""`</span> necessary).</span>
<span id="cb82-826"><a href="#cb82-826" aria-hidden="true" tabindex="-1"></a>For example, to select only the <span class="in">`mouse_strain`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`m_dose`</span> data frame:</span>
<span id="cb82-827"><a href="#cb82-827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-830"><a href="#cb82-830" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-831"><a href="#cb82-831" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-832"><a href="#cb82-832" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain, initial_weight_g, weight_lost_g)</span>
<span id="cb82-833"><a href="#cb82-833" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-834"><a href="#cb82-834" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-835"><a href="#cb82-835" aria-hidden="true" tabindex="-1"></a>We can see that all the other columns have been removed from the data frame.</span>
<span id="cb82-836"><a href="#cb82-836" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-837"><a href="#cb82-837" aria-hidden="true" tabindex="-1"></a>If you want to keep all columns except for a few, you can use <span class="in">`-`</span> to drop columns.</span>
<span id="cb82-838"><a href="#cb82-838" aria-hidden="true" tabindex="-1"></a>For example, to keep all columns except for <span class="in">`cage_number`</span> and <span class="in">`sex`</span>:</span>
<span id="cb82-839"><a href="#cb82-839" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-842"><a href="#cb82-842" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-843"><a href="#cb82-843" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-844"><a href="#cb82-844" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>cage_number, <span class="sc">-</span>sex)</span>
<span id="cb82-845"><a href="#cb82-845" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-846"><a href="#cb82-846" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-847"><a href="#cb82-847" aria-hidden="true" tabindex="-1"></a>There are also some helper functions that can be used to select columns based on their names :</span>
<span id="cb82-848"><a href="#cb82-848" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-849"><a href="#cb82-849" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb82-850"><a href="#cb82-850" aria-hidden="true" tabindex="-1"></a>| Function        | Description                                       | Example                                         |</span>
<span id="cb82-851"><a href="#cb82-851" aria-hidden="true" tabindex="-1"></a>+=================+===================================================+=================================================+</span>
<span id="cb82-852"><a href="#cb82-852" aria-hidden="true" tabindex="-1"></a>| <span class="in">`starts_with()`</span> | select column(s) that start with a certain string | select all columns starting with the letter i   |</span>
<span id="cb82-853"><a href="#cb82-853" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb82-854"><a href="#cb82-854" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(starts_with("i"))`</span>                      |</span>
<span id="cb82-855"><a href="#cb82-855" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb82-856"><a href="#cb82-856" aria-hidden="true" tabindex="-1"></a>| <span class="in">`ends_with()`</span>   | select column(s) that end with a certain string   | select all columns ending with <span class="sc">\_</span>g              |</span>
<span id="cb82-857"><a href="#cb82-857" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb82-858"><a href="#cb82-858" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(ends_with("_g"))`</span>                       |</span>
<span id="cb82-859"><a href="#cb82-859" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb82-860"><a href="#cb82-860" aria-hidden="true" tabindex="-1"></a>| <span class="in">`contains()`</span>    | select column(s) that contain a certain string    | select all columns containing the word 'weight' |</span>
<span id="cb82-861"><a href="#cb82-861" aria-hidden="true" tabindex="-1"></a>|                 |                                                   |                                                 |</span>
<span id="cb82-862"><a href="#cb82-862" aria-hidden="true" tabindex="-1"></a>|                 |                                                   | <span class="in">`select(contains("weight"))`</span>                    |</span>
<span id="cb82-863"><a href="#cb82-863" aria-hidden="true" tabindex="-1"></a>+-----------------+---------------------------------------------------+-------------------------------------------------+</span>
<span id="cb82-864"><a href="#cb82-864" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-865"><a href="#cb82-865" aria-hidden="true" tabindex="-1"></a>: There are several helper functions that can be used with the select function</span>
<span id="cb82-866"><a href="#cb82-866" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-867"><a href="#cb82-867" aria-hidden="true" tabindex="-1"></a>You need to use quotation marks around the arguments in these helper functions, as they aren't full column names, just strings of characters.</span>
<span id="cb82-868"><a href="#cb82-868" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-869"><a href="#cb82-869" aria-hidden="true" tabindex="-1"></a>Try using these helper functions to select columns from the <span class="in">`m_dose`</span> data frame!</span>
<span id="cb82-870"><a href="#cb82-870" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-871"><a href="#cb82-871" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Reordering columns"}</span>
<span id="cb82-872"><a href="#cb82-872" aria-hidden="true" tabindex="-1"></a><span class="al">![Relocate allows you to move columns around](images/S02F05_reorder.png)</span></span>
<span id="cb82-873"><a href="#cb82-873" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-874"><a href="#cb82-874" aria-hidden="true" tabindex="-1"></a>We can reorder columns using the <span class="in">`relocate()`</span> function, which works similarly to <span class="in">`select()`</span> (except it just moves columns around rather than dropping/keeping them).</span>
<span id="cb82-875"><a href="#cb82-875" aria-hidden="true" tabindex="-1"></a>For example, to move the <span class="in">`sex`</span> column to before the <span class="in">`cage_number`</span> column:</span>
<span id="cb82-876"><a href="#cb82-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-879"><a href="#cb82-879" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-880"><a href="#cb82-880" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-881"><a href="#cb82-881" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first the name of the column to move, then where it should go</span></span>
<span id="cb82-882"><a href="#cb82-882" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(sex, <span class="at">.before =</span> cage_number)</span>
<span id="cb82-883"><a href="#cb82-883" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-884"><a href="#cb82-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-885"><a href="#cb82-885" aria-hidden="true" tabindex="-1"></a>Without a specific position <span class="in">`( .before / .after)`</span>, this function will place the chosen column(s) as the first / left-most columns.</span>
<span id="cb82-886"><a href="#cb82-886" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-887"><a href="#cb82-887" aria-hidden="true" tabindex="-1"></a>Two further useful helper functions for <span class="in">`relocate()`</span> are the <span class="in">`everything()`</span> and <span class="in">`last_col()`</span> functions, which can be used to move columns to the start/end of the data frame.</span>
<span id="cb82-888"><a href="#cb82-888" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-891"><a href="#cb82-891" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-892"><a href="#cb82-892" aria-hidden="true" tabindex="-1"></a><span class="co"># move id_num to the front</span></span>
<span id="cb82-893"><a href="#cb82-893" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-894"><a href="#cb82-894" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(id_num, <span class="at">.before =</span> <span class="fu">everything</span>()) <span class="co"># don't forget the brackets</span></span>
<span id="cb82-895"><a href="#cb82-895" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-896"><a href="#cb82-896" aria-hidden="true" tabindex="-1"></a><span class="co"># move mouse_strain to the end</span></span>
<span id="cb82-897"><a href="#cb82-897" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-898"><a href="#cb82-898" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(mouse_strain, <span class="at">.after =</span> <span class="fu">last_col</span>())</span>
<span id="cb82-899"><a href="#cb82-899" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-900"><a href="#cb82-900" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-901"><a href="#cb82-901" aria-hidden="true" tabindex="-1"></a>Re-ordering columns isn't necessary, but it makes it easier to see the data you're most interested in within the console (since often not all of the columns will fit on the screen at once).</span>
<span id="cb82-902"><a href="#cb82-902" aria-hidden="true" tabindex="-1"></a>For example, if we are doing a lot of computation on the <span class="in">`initial_weight_g`</span> column, we'd probably like to have that near the start so we can easily check it.</span>
<span id="cb82-903"><a href="#cb82-903" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-904"><a href="#cb82-904" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-905"><a href="#cb82-905" aria-hidden="true" tabindex="-1"></a>Note that the output of the <span class="in">`select()`</span> function is a new data frame, even if you only select a single column:</span>
<span id="cb82-906"><a href="#cb82-906" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-909"><a href="#cb82-909" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-910"><a href="#cb82-910" aria-hidden="true" tabindex="-1"></a><span class="co"># select the mouse_strain column</span></span>
<span id="cb82-911"><a href="#cb82-911" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-912"><a href="#cb82-912" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb82-913"><a href="#cb82-913" aria-hidden="true" tabindex="-1"></a>  <span class="co"># recall from chapter 1 that class() tells us the type of an object</span></span>
<span id="cb82-914"><a href="#cb82-914" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span>
<span id="cb82-915"><a href="#cb82-915" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-916"><a href="#cb82-916" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-917"><a href="#cb82-917" aria-hidden="true" tabindex="-1"></a>Sometimes, we instead want to get the values of a column as a vector.</span>
<span id="cb82-918"><a href="#cb82-918" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-919"><a href="#cb82-919" aria-hidden="true" tabindex="-1"></a><span class="al">![Pull allows you to pull a column out of a data frame as a vector](images/S02F06_pull.png)</span></span>
<span id="cb82-920"><a href="#cb82-920" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-921"><a href="#cb82-921" aria-hidden="true" tabindex="-1"></a>We can do this by using the <span class="in">`pull()`</span> function, which extracts a single column from a data frame as a vector:</span>
<span id="cb82-922"><a href="#cb82-922" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-925"><a href="#cb82-925" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-926"><a href="#cb82-926" aria-hidden="true" tabindex="-1"></a><span class="co"># get the mouse_strain column as a vector</span></span>
<span id="cb82-927"><a href="#cb82-927" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-928"><a href="#cb82-928" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(mouse_strain) <span class="sc">%&gt;%</span></span>
<span id="cb82-929"><a href="#cb82-929" aria-hidden="true" tabindex="-1"></a>  <span class="fu">class</span>()</span>
<span id="cb82-930"><a href="#cb82-930" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-931"><a href="#cb82-931" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-932"><a href="#cb82-932" aria-hidden="true" tabindex="-1"></a>We can see that the class of the output is now a vector, rather than a data frame.</span>
<span id="cb82-933"><a href="#cb82-933" aria-hidden="true" tabindex="-1"></a>This is important because some functions only accept vectors, not data frames, like <span class="in">`mean()`</span> for example:</span>
<span id="cb82-934"><a href="#cb82-934" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-937"><a href="#cb82-937" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-938"><a href="#cb82-938" aria-hidden="true" tabindex="-1"></a><span class="co"># this will give an error</span></span>
<span id="cb82-939"><a href="#cb82-939" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span> <span class="fu">select</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-940"><a href="#cb82-940" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-941"><a href="#cb82-941" aria-hidden="true" tabindex="-1"></a><span class="co"># this will work</span></span>
<span id="cb82-942"><a href="#cb82-942" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span> <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span> <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-943"><a href="#cb82-943" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-944"><a href="#cb82-944" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-945"><a href="#cb82-945" aria-hidden="true" tabindex="-1"></a>Note how both times we used <span class="in">`na.rm = TRUE`</span> to remove missing values before calculating the mean.</span>
<span id="cb82-946"><a href="#cb82-946" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-947"><a href="#cb82-947" aria-hidden="true" tabindex="-1"></a>You might remember that we used the <span class="in">`$`</span> operator in the previous chapter to extract a single column from a data frame, so why use <span class="in">`pull()`</span> instead?</span>
<span id="cb82-948"><a href="#cb82-948" aria-hidden="true" tabindex="-1"></a>The main reason is that <span class="in">`pull()`</span> works within a chain of pipes, whereas <span class="in">`$`</span> doesn't.</span>
<span id="cb82-949"><a href="#cb82-949" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-950"><a href="#cb82-950" aria-hidden="true" tabindex="-1"></a>For example, let's say we want to know the average initial weight of mice that lost at least 4g.</span>
<span id="cb82-951"><a href="#cb82-951" aria-hidden="true" tabindex="-1"></a>We can do this by chaining <span class="in">`filter()`</span> and <span class="in">`pull()`</span> together:</span>
<span id="cb82-952"><a href="#cb82-952" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-955"><a href="#cb82-955" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb82-956"><a href="#cb82-956" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb82-957"><a href="#cb82-957" aria-hidden="true" tabindex="-1"></a>  <span class="co"># filter to mice that lost at least 4g</span></span>
<span id="cb82-958"><a href="#cb82-958" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(weight_lost_g <span class="sc">&gt;=</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb82-959"><a href="#cb82-959" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the initial_weight_g column as a vector</span></span>
<span id="cb82-960"><a href="#cb82-960" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(initial_weight_g) <span class="sc">%&gt;%</span></span>
<span id="cb82-961"><a href="#cb82-961" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate mean, removing NA values</span></span>
<span id="cb82-962"><a href="#cb82-962" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-963"><a href="#cb82-963" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb82-964"><a href="#cb82-964" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-965"><a href="#cb82-965" aria-hidden="true" tabindex="-1"></a>::::::::::::::::::::: {.callout-important title="Practice exercises"}</span>
<span id="cb82-966"><a href="#cb82-966" aria-hidden="true" tabindex="-1"></a>Try these practice questions to test your understanding</span>
<span id="cb82-967"><a href="#cb82-967" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-968"><a href="#cb82-968" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-969"><a href="#cb82-969" aria-hidden="true" tabindex="-1"></a>1<span class="sc">\.</span></span>
<span id="cb82-970"><a href="#cb82-970" aria-hidden="true" tabindex="-1"></a>Which line of code would NOT be a valid way to select only the <span class="in">`drug_dose_g`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`m_dose`</span> data frame?</span>
<span id="cb82-971"><a href="#cb82-971" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-972"><a href="#cb82-972" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-973"><a href="#cb82-973" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-974"><a href="#cb82-974" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% select(drug_dose_g, initial_weight_g, weight_lost_g)`</span></span>
<span id="cb82-975"><a href="#cb82-975" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-976"><a href="#cb82-976" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-977"><a href="#cb82-977" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-978"><a href="#cb82-978" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% select(contains("g"))`</span></span>
<span id="cb82-979"><a href="#cb82-979" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-980"><a href="#cb82-980" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-981"><a href="#cb82-981" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-982"><a href="#cb82-982" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% select(ends_with("_g"))`</span></span>
<span id="cb82-983"><a href="#cb82-983" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-984"><a href="#cb82-984" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-985"><a href="#cb82-985" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-986"><a href="#cb82-986" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% select(-cage_number, -tail_length_mm, -id_num, -mouse_strain, -sex, -replicate)`</span></span>
<span id="cb82-987"><a href="#cb82-987" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-988"><a href="#cb82-988" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-989"><a href="#cb82-989" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-990"><a href="#cb82-990" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-991"><a href="#cb82-991" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-992"><a href="#cb82-992" aria-hidden="true" tabindex="-1"></a>2<span class="sc">\.</span></span>
<span id="cb82-993"><a href="#cb82-993" aria-hidden="true" tabindex="-1"></a>How would I extract the <span class="in">`initial_weight_g`</span> column from the <span class="in">`m_dose`</span> data frame as a vector?</span>
<span id="cb82-994"><a href="#cb82-994" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-995"><a href="#cb82-995" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-996"><a href="#cb82-996" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-997"><a href="#cb82-997" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% filter(initial_weight_g)`</span></span>
<span id="cb82-998"><a href="#cb82-998" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-999"><a href="#cb82-999" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1000"><a href="#cb82-1000" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-1001"><a href="#cb82-1001" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% $initial_weight_g`</span></span>
<span id="cb82-1002"><a href="#cb82-1002" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1003"><a href="#cb82-1003" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1004"><a href="#cb82-1004" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-1005"><a href="#cb82-1005" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% select(initial_weight_g)`</span></span>
<span id="cb82-1006"><a href="#cb82-1006" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1007"><a href="#cb82-1007" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1008"><a href="#cb82-1008" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-1009"><a href="#cb82-1009" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% pull(initial_weight_g)`</span></span>
<span id="cb82-1010"><a href="#cb82-1010" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1011"><a href="#cb82-1011" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-1012"><a href="#cb82-1012" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-1013"><a href="#cb82-1013" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1014"><a href="#cb82-1014" aria-hidden="true" tabindex="-1"></a>:::::::: question</span>
<span id="cb82-1015"><a href="#cb82-1015" aria-hidden="true" tabindex="-1"></a>3<span class="sc">\.</span></span>
<span id="cb82-1016"><a href="#cb82-1016" aria-hidden="true" tabindex="-1"></a>How would you move the <span class="in">`sex`</span> column to the end of the <span class="in">`m_dose`</span> data frame?</span>
<span id="cb82-1017"><a href="#cb82-1017" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1018"><a href="#cb82-1018" aria-hidden="true" tabindex="-1"></a>::::::: choices</span>
<span id="cb82-1019"><a href="#cb82-1019" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-1020"><a href="#cb82-1020" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% relocate(sex)`</span></span>
<span id="cb82-1021"><a href="#cb82-1021" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1022"><a href="#cb82-1022" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1023"><a href="#cb82-1023" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-1024"><a href="#cb82-1024" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% relocate(sex, .after = last_col)`</span></span>
<span id="cb82-1025"><a href="#cb82-1025" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1026"><a href="#cb82-1026" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1027"><a href="#cb82-1027" aria-hidden="true" tabindex="-1"></a>::: {.choice .correct-choice}</span>
<span id="cb82-1028"><a href="#cb82-1028" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% relocate(sex, .after = last_col())`</span></span>
<span id="cb82-1029"><a href="#cb82-1029" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1030"><a href="#cb82-1030" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1031"><a href="#cb82-1031" aria-hidden="true" tabindex="-1"></a>::: choice</span>
<span id="cb82-1032"><a href="#cb82-1032" aria-hidden="true" tabindex="-1"></a><span class="in">`m_dose %&gt;% reorder(sex, .after = last_col())`</span></span>
<span id="cb82-1033"><a href="#cb82-1033" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1034"><a href="#cb82-1034" aria-hidden="true" tabindex="-1"></a>:::::::</span>
<span id="cb82-1035"><a href="#cb82-1035" aria-hidden="true" tabindex="-1"></a>::::::::</span>
<span id="cb82-1036"><a href="#cb82-1036" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1037"><a href="#cb82-1037" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb82-1038"><a href="#cb82-1038" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1039"><a href="#cb82-1039" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb82-1040"><a href="#cb82-1040" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1041"><a href="#cb82-1041" aria-hidden="true" tabindex="-1"></a>&lt;p&gt;</span>
<span id="cb82-1042"><a href="#cb82-1042" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1043"><a href="#cb82-1043" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The line of code that would NOT be a valid way to select the <span class="in">`drug_dose_g`</span>, <span class="in">`initial_weight_g`</span>, and <span class="in">`weight_lost_g`</span> columns from the <span class="in">`m_dose`</span> data frame is <span class="in">`m_dose %&gt;% select(contains("g"))`</span>. This line of code would select all columns that contain the letter 'g', which would include columns like <span class="in">`cage_number`</span> and <span class="in">`tail_length_mm`</span>. We need to specify either <span class="in">`ends_with("g")`</span> or <span class="in">`contains("_g")`</span> to only get those with <span class="in">`_g`</span> at the end. The other options are valid ways to select the specified columns, although some are more efficient than others!</span>
<span id="cb82-1044"><a href="#cb82-1044" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The correct way to extract the <span class="in">`initial_weight_g`</span> column from the <span class="in">`m_dose`</span> data frame as a vector is <span class="in">`m_dose %&gt;% pull(initial_weight_g)`</span>. The <span class="in">`pull()`</span> function is used to extract a single column from a data frame as a vector. The other options are incorrect because <span class="in">`filter()`</span> is used to subset rows, <span class="in">`$`</span> is not used in a pipe chain, and <span class="in">`select()`</span> is outputs a data frame, not extract them as vectors.</span>
<span id="cb82-1045"><a href="#cb82-1045" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The correct way to move the <span class="in">`sex`</span> column to the end of the <span class="in">`m_dose`</span> data frame is using the <span class="in">`relocate()`</span> function like this: <span class="in">`m_dose %&gt;% relocate(sex, .after = last_col())`</span>. The <span class="in">`last_col()`</span> function is used to refer to the last column in the data frame. The other options are incorrect because <span class="in">`reorder()`</span> is not a valid function, and you need to remember to include the brackets <span class="in">`()`</span> when using <span class="in">`last_col()`</span>.</span>
<span id="cb82-1046"><a href="#cb82-1046" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1047"><a href="#cb82-1047" aria-hidden="true" tabindex="-1"></a>&lt;/p&gt;</span>
<span id="cb82-1048"><a href="#cb82-1048" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1049"><a href="#cb82-1049" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
<span id="cb82-1050"><a href="#cb82-1050" aria-hidden="true" tabindex="-1"></a>:::::::::::::::::::::</span>
<span id="cb82-1051"><a href="#cb82-1051" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1052"><a href="#cb82-1052" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb82-1053"><a href="#cb82-1053" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1054"><a href="#cb82-1054" aria-hidden="true" tabindex="-1"></a>Here's what we've covered in this chapter:</span>
<span id="cb82-1055"><a href="#cb82-1055" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1056"><a href="#cb82-1056" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The pipe operator <span class="in">`%&gt;%`</span> and how we can use it to chain together multiple function calls, making our code more readable and easier to understand.</span>
<span id="cb82-1057"><a href="#cb82-1057" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1058"><a href="#cb82-1058" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The basic dplyr verbs <span class="in">`arrange()`</span>, <span class="in">`filter()`</span> and <span class="in">`select()`</span></span>
<span id="cb82-1059"><a href="#cb82-1059" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1060"><a href="#cb82-1060" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Why does data need to be tidy anyway?"}</span>
<span id="cb82-1061"><a href="#cb82-1061" aria-hidden="true" tabindex="-1"></a>In this chapter, we've been focusing on making our data 'tidy': that is, structured in a consistent way that makes it easy to work with.</span>
<span id="cb82-1062"><a href="#cb82-1062" aria-hidden="true" tabindex="-1"></a>A nice visual illustration of tidy data and its importance can be <span class="co">[</span><span class="ot">found here</span><span class="co">](https://allisonhorst.com/other-r-fun)</span>.</span>
<span id="cb82-1063"><a href="#cb82-1063" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb82-1064"><a href="#cb82-1064" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1065"><a href="#cb82-1065" aria-hidden="true" tabindex="-1"></a><span class="fu">### Practice questions</span></span>
<span id="cb82-1066"><a href="#cb82-1066" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1067"><a href="#cb82-1067" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>What is the purpose of the pipe operator <span class="in">`%&gt;%`</span>?</span>
<span id="cb82-1068"><a href="#cb82-1068" aria-hidden="true" tabindex="-1"></a>    Keeping this in mind, re-write the following code to use the pipe.</span>
<span id="cb82-1069"><a href="#cb82-1069" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1070"><a href="#cb82-1070" aria-hidden="true" tabindex="-1"></a>    a.  <span class="in">`round(mean(c(1, 2, 3, 4, 5)))`</span></span>
<span id="cb82-1071"><a href="#cb82-1071" aria-hidden="true" tabindex="-1"></a>    b.  <span class="in">`print(as.character(1 + 10))`</span></span>
<span id="cb82-1072"><a href="#cb82-1072" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1073"><a href="#cb82-1073" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>What would be the result of evaluating the following expressions?</span>
<span id="cb82-1074"><a href="#cb82-1074" aria-hidden="true" tabindex="-1"></a>    You don't need to know these off the top of your head, use R to help!</span>
<span id="cb82-1075"><a href="#cb82-1075" aria-hidden="true" tabindex="-1"></a>    (Hint: some expressions might give an error. Try to think about why)</span>
<span id="cb82-1076"><a href="#cb82-1076" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1077"><a href="#cb82-1077" aria-hidden="true" tabindex="-1"></a>    a.  <span class="in">`m_dose %&gt;% filter(weight_lost_g &gt; 10)`</span></span>
<span id="cb82-1078"><a href="#cb82-1078" aria-hidden="true" tabindex="-1"></a>    b.  <span class="in">`m_dose %&gt;% select(tail_length_mm, weight_lost_g)`</span></span>
<span id="cb82-1079"><a href="#cb82-1079" aria-hidden="true" tabindex="-1"></a>    c.  <span class="in">`m_dose %&gt;% arrange(tail_length_mm)`</span></span>
<span id="cb82-1080"><a href="#cb82-1080" aria-hidden="true" tabindex="-1"></a>    d.  <span class="in">`m_dose %&gt;% filter(initial_Weight_g &gt; 10) %&gt;% arrange(mouse_strain)`</span></span>
<span id="cb82-1081"><a href="#cb82-1081" aria-hidden="true" tabindex="-1"></a>    e.  <span class="in">`m_dose %&gt;% relocate(mouse_strain, .after = cage_number)`</span></span>
<span id="cb82-1082"><a href="#cb82-1082" aria-hidden="true" tabindex="-1"></a>    f.  <span class="in">`m_dose %&gt;% pull(weight_lost_g)`</span></span>
<span id="cb82-1083"><a href="#cb82-1083" aria-hidden="true" tabindex="-1"></a>    g.  <span class="in">`m_dose %&gt;% filter(!is.na(weight_lost_g))`</span></span>
<span id="cb82-1084"><a href="#cb82-1084" aria-hidden="true" tabindex="-1"></a>    h.  <span class="in">`m_dose %&gt;% replace_na(list(weight_lost_g = 0))`</span></span>
<span id="cb82-1085"><a href="#cb82-1085" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1086"><a href="#cb82-1086" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>What is a missing value in R?</span>
<span id="cb82-1087"><a href="#cb82-1087" aria-hidden="true" tabindex="-1"></a>    What are two ways to deal with missing values in a data frame?</span>
<span id="cb82-1088"><a href="#cb82-1088" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1089"><a href="#cb82-1089" aria-hidden="true" tabindex="-1"></a>&lt;details&gt;</span>
<span id="cb82-1090"><a href="#cb82-1090" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1091"><a href="#cb82-1091" aria-hidden="true" tabindex="-1"></a>&lt;summary&gt;Solutions&lt;/summary&gt;</span>
<span id="cb82-1092"><a href="#cb82-1092" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1093"><a href="#cb82-1093" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The pipe operator <span class="in">`%&gt;%`</span> is used to chain together multiple function calls, passing the result of one function to the next.</span>
<span id="cb82-1094"><a href="#cb82-1094" aria-hidden="true" tabindex="-1"></a>    Here's how you could re-write the code to use the pipe:</span>
<span id="cb82-1095"><a href="#cb82-1095" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1096"><a href="#cb82-1096" aria-hidden="true" tabindex="-1"></a>    a.  <span class="in">`c(1, 2, 3, 4, 5) %&gt;% mean() %&gt;% round()`</span></span>
<span id="cb82-1097"><a href="#cb82-1097" aria-hidden="true" tabindex="-1"></a>    b.  <span class="in">`as.character(1 + 10) %&gt;% print()`</span></span>
<span id="cb82-1098"><a href="#cb82-1098" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1099"><a href="#cb82-1099" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The result of evaluating the expressions would be:</span>
<span id="cb82-1100"><a href="#cb82-1100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1101"><a href="#cb82-1101" aria-hidden="true" tabindex="-1"></a>    a.  A data frame containing only the rows where <span class="in">`weight_lost_g`</span> is greater than 10.</span>
<span id="cb82-1102"><a href="#cb82-1102" aria-hidden="true" tabindex="-1"></a>    b.  A data frame containing only the <span class="in">`tail_length_mm`</span> and <span class="in">`weight_lost_g`</span> columns.</span>
<span id="cb82-1103"><a href="#cb82-1103" aria-hidden="true" tabindex="-1"></a>    c.  A data frame sorted by <span class="in">`tail_length_mm`</span>, in ascending order.</span>
<span id="cb82-1104"><a href="#cb82-1104" aria-hidden="true" tabindex="-1"></a>    d.  An error because <span class="in">`initial_Weight_g`</span> is not a column in the data frame.</span>
<span id="cb82-1105"><a href="#cb82-1105" aria-hidden="true" tabindex="-1"></a>    e.  A data frame with the <span class="in">`mouse_strain`</span> column moved to be after the <span class="in">`cage_number`</span> column.</span>
<span id="cb82-1106"><a href="#cb82-1106" aria-hidden="true" tabindex="-1"></a>    f.  A **vector** containing the values of the <span class="in">`weight_lost_g`</span> column.</span>
<span id="cb82-1107"><a href="#cb82-1107" aria-hidden="true" tabindex="-1"></a>    g.  A data frame containing only the rows where <span class="in">`weight_lost_g`</span> is not <span class="in">`NA`</span>.</span>
<span id="cb82-1108"><a href="#cb82-1108" aria-hidden="true" tabindex="-1"></a>    h.  A data frame with missing values in the <span class="in">`weight_lost_g`</span> column replaced with 0.</span>
<span id="cb82-1109"><a href="#cb82-1109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1110"><a href="#cb82-1110" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>A missing value in R is represented by <span class="in">`NA`</span>.</span>
<span id="cb82-1111"><a href="#cb82-1111" aria-hidden="true" tabindex="-1"></a>    Two ways to deal with missing values in a data frame are to remove them using <span class="in">`filter(!is.na(column_name))`</span> or to replace them with a specific value using <span class="in">`replace_na(list(column_name = value))`</span>.</span>
<span id="cb82-1112"><a href="#cb82-1112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-1113"><a href="#cb82-1113" aria-hidden="true" tabindex="-1"></a>&lt;/details&gt;</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>